---
title: "Monocle3"
output: html_document
date: "2025-03-24"
---

https://www.youtube.com/watch?v=YQqVsXdwFNU&t=206s
https://github.com/cole-trapnell-lab/monocle3
https://cole-trapnell-lab.github.io/monocle3/
https://cole-trapnell-lab.github.io/monocle3/docs/installation/

# install
```{r}
# BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
#                        'limma', 'lme4', 'S4Vectors', 'SingleCellExperiment',
#                        'SummarizedExperiment', 'batchelor', 'HDF5Array',
#                        'terra', 'ggrastr'))
# install.packages("devtools")
devtools::install_github('cole-trapnell-lab/monocle3')
library(monocle3)
```


# Lecture 1
```{r}
library(monocle3)
library(tidyverse)
library(patchwork)
set.seed(1234)

#1. Generate a cell_data_set from 10X output (umi_cutoff = 100)
cds1 <- load_mm_data(mat_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF1Lin/matrix.mtx", 
                    feature_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF1Lin/features.tsv", 
                    cell_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF1Lin/barcodes.tsv")

colData(cds1)
colData(cds1)$samples <- paste("IPF1")

cds2 <- load_mm_data(mat_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF2Lin/matrix.mtx", 
                     feature_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF2Lin/features.tsv", 
                     cell_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF2Lin/barcodes.tsv")

colData(cds2)$samples <- paste("IPF2")

cds3 <- load_mm_data(mat_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF3Lin/matrix.mtx", 
                     feature_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF3Lin/features.tsv", 
                     cell_anno_path = "../Desktop/Video_Tutorials/Data/GSE132771/IPF3Lin/barcodes.tsv")

colData(cds3)$samples <- paste("IPF3")

cds <- combine_cds(list(cds1, cds2, cds3))

rm (cds1, cds2, cds3)

fData(cds)
names(fData(cds)) <- "gene_short_name"
fData(cds)

# 2. Pre-process the data
#    normalize the data & Principal Components Analysis
#    specify the number of principal components you want Monocle to compute
cds <- preprocess_cds(cds, num_dim = 100) 

# check if we used enough PCs to capture most of the variation in gene expression across all the cells
plot_pc_variance_explained(cds)

# 3. Reduce dimensionality: UMAP by default, t-SNE 
cds <- reduce_dimension(cds)

# 4. Visualize the cells and Check batch effects
colData(cds)
plot_cells(cds, color_cells_by="samples", group_label_size = 6)

# 5. Remove batch effects
cds <- align_cds(cds, num_dim = 100, alignment_group = "sample")
cds <- reduce_dimension(cds)
plot_cells(cds, color_cells_by="samples", group_label_size = 6)

# 6. Group cells into clusters
cds <- cluster_cells(cds, resolution=1e-5)
plot_cells(cds, group_label_size = 6)

# 7. Marker genes
marker_test_res <- top_markers(cds, group_cells_by="cluster", verbose = T)

top_specific_markers <- marker_test_res %>% 
  filter(fraction_expressing >= 0.10) %>%
  group_by(cell_group) %>%
  top_n(3, marker_score) # n=3

top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))

plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="cluster",
                    ordering_type="maximal_on_diag",
                    max.size=3)

P1 <- plot_cells(cds, color_cells_by="cluster", group_cells_by="cluster", group_label_size = 6)
P2 <- plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="cluster",
                    ordering_type="maximal_on_diag",
                    max.size=3)
P1|P2

# "1"="Alveolar Fibroblasts", "2"="Pericytes","3"="SMCs","4"="Airway Fibroblasts"
# "5"="Mesothelial", "6"="IM","7"="AT2","8"="Basal cells"

# 8. Annotate your cells according to the cell types
clusters(cds)

colData(cds)$assigned_cell_type <- as.character(clusters(cds))

colData(cds)

colData(cds)$assigned_cell_type <- dplyr::recode(colData(cds)$assigned_cell_type,
   "1"="Alveolar Fibroblasts", "2"="Pericytes","3"="SMCs","4"="Airway Fibroblasts",           
    "5"="Mesothelial", "6"="IM","7"="AT2","8"="Basal cells")

colData(cds)

plot_cells(cds, group_cells_by="cluster", color_cells_by="assigned_cell_type",
           group_label_size = 6)

saveRDS(cds, file="../Desktop/Video_Tutorials/Data/GSE132771/cds.RDS")

save_monocle_objects(cds=cds, directory_path='../Desktop/Video_Tutorials/Data/')
cds <- load_monocle_objects(directory_path='my_cds_objects')
```

# Lecture 2

```{r}
# 1. Load the analysed data
cds <- readRDS("../Desktop/Video_Tutorials/Data/GSE132771/cds.RDS")
plot_cells(cds, group_cells_by="cluster", color_cells_by="assigned_cell_type",
           group_label_size = 6)

# 2. subset SMCs, Pericytes and fibroblasts
cds_subset <- choose_cells(cds)
cds_subset <- reduce_dimension(cds_subset)
cds_subset <- cluster_cells(cds_subset, resolution = 1e-5)
plot_cells(cds_subset, color_cells_by="assigned_cell_type", group_label_size = 6)

saveRDS(cds_subset, file="../Desktop/Video_Tutorials/Data/GSE132771/cds_subset.RDS")

plot_cells(cds_subset, genes=c("PDGFRB", "ACTA2", "RGS5", "CTHRC1"))
plot_cells(cds_subset, genes=c("COL1A1", "COL3A1", "CTHRC1", "POSTN"))
plot_cells(cds_subset, genes=c("CD34", "PI16", "SCARA5", "MFAP5"))

# 3. Constructing Single Cell Trajectories 
cds_subset <- learn_graph(cds_subset)

plot_cells(cds_subset, color_cells_by = "assigned_cell_type",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           group_label_size = 6)

cds_subset <- order_cells(cds_subset)

plot_cells(cds_subset, color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size= 3)
```


# Lecture 3
```{r}
library(Seurat)
library(tidyverse)
library(SeuratWrappers)
library(monocle3)

# 1. Analyze scRNA-seq data using Seurat------
# Read
IPF_I<- Read10X(data.dir = "../Desktop/Video_Tutorials/Data/GSE132771/IPF1Lin/")
IPF_II<- Read10X(data.dir = "../Desktop/Video_Tutorials/Data/GSE132771/IPF2Lin/")
IPF_III<- Read10X(data.dir = "../Desktop/Video_Tutorials/Data/GSE132771/IPF3Lin/")

# Create Seurat Object
IPF_I <- CreateSeuratObject(counts = IPF_I, project = "IPF_I", 
                            min.cells = 3, min.features = 200) 
IPF_I <- PercentageFeatureSet(IPF_I, pattern = "^MT-", col.name = "percent.mt")

IPF_II <- CreateSeuratObject(counts = IPF_II, project = "IPF_II", 
                             min.cells = 3, min.features = 200)
IPF_II <- PercentageFeatureSet(IPF_II, pattern = "^MT-", col.name = "percent.mt")

IPF_III <- CreateSeuratObject(counts = IPF_III, project = "IPF_III", 
                              min.cells = 3, min.features = 200)
IPF_III <- PercentageFeatureSet(IPF_III,pattern = "^MT-",col.name = "percent.mt")

IPF <- merge(IPF_I, y = c(IPF_II, IPF_III))
VlnPlot(IPF, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
IPF <- subset(IPF, nFeature_RNA<5000 & nCount_RNA<20000 & percent.mt<10)

View(IPF@meta.data)

IPF.list <- SplitObject(IPF, split.by = "orig.ident")

# normalize and identify variable features for each dataset independently
IPF.list <- lapply(X = IPF.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = IPF.list)
IPF.anchors <- FindIntegrationAnchors(object.list = IPF.list, 
                                      anchor.features = features)

# creates an 'integrated' data assay
IPF.combined <- IntegrateData(anchorset = IPF.anchors)

# specify that we will perform downstream analysis on the integrated data
DefaultAssay(IPF.combined) <- "integrated"

# Run the standard workflow for cell clustering
IPF.combined <- ScaleData(IPF.combined, verbose = FALSE)
IPF.combined <- RunPCA(IPF.combined, npcs = 30, verbose = FALSE)
IPF.combined <- RunUMAP(IPF.combined, reduction = "pca", dims = 1:30)
IPF.combined <- FindNeighbors(IPF.combined, reduction = "pca", dims = 1:30)
IPF.combined <- FindClusters(IPF.combined, resolution = 0.5)

# Visualization
DimPlot(IPF.combined, reduction = "umap", label = TRUE)

# four markers
FeaturePlot(IPF.combined, features=c("COL1A2", "LUM", "PDGFRA","PDGFRB"),
            label = T, cols = c('lightgrey', 'blue'))
FeaturePlot(IPF.combined, features=c("EPCAM", "SFTPC", "AGER", "SCGB1A1"),
            label = T, cols = c('lightgrey', 'blue'))
FeaturePlot(IPF.combined, features=c("CLDN5", "CCL21", "PECAM1", "EMCN"),
            label = T, cols = c('lightgrey', 'blue'))
FeaturePlot(IPF.combined, features=c("PTPRC", "CD52", "AIF1", "TRBC2"),
            label = T,cols = c('lightgrey', 'blue'))
FeaturePlot(IPF.combined, features=c("MSLN", "CALB2", "HP", "PRG4"),
            label = T, cols = c('lightgrey', 'blue'))

IPF_Mesenchymal <- subset(x = IPF.combined, idents = c("0", "1", "2", 
                          "3", "4", "5", "6", "7", "8", "9"))
DimPlot(IPF_Mesenchymal, reduction = "umap", label = TRUE)

# Run the standard workflow for cell clustering
IPF_Mesenchymal <- ScaleData(IPF_Mesenchymal, verbose = FALSE)
IPF_Mesenchymal <- RunPCA(IPF_Mesenchymal, npcs = 30, verbose = FALSE)
IPF_Mesenchymal <- RunUMAP(IPF_Mesenchymal, reduction = "pca", dims = 1:30)
IPF_Mesenchymal <- FindNeighbors(IPF_Mesenchymal, reduction = "pca", dims = 1:30)
IPF_Mesenchymal <- FindClusters(IPF_Mesenchymal, resolution = 0.3)

# Visualization
DimPlot(IPF_Mesenchymal, reduction = "umap", label = TRUE)

IPF_Mesenchymal <- subset(x = IPF_Mesenchymal, idents = c("0", "1", "2", 
                                                       "3", "4", "5", "6"))

FeaturePlot(IPF_Mesenchymal, features=c("RGS5", "ACTA2", "LUM", "ACTG2"),
            label = T, cols = c('lightgrey', 'blue'))
FeaturePlot(IPF_Mesenchymal, features=c("CTHRC1", "COL1A1", "COL3A1", "POSTN"),
            label = T, cols = c('lightgrey', 'blue'))

IPF_Mesenchymal <- RenameIdents(IPF_Mesenchymal,
       `0` = "Pericytes", `4` = "Pericytes", `1` = "SMCs",
       `3` = "MyoF", `2` = "AlvF", `5` = "AlvF", `6` = "AirwayF")

DimPlot(IPF_Mesenchymal , reduction = "umap", label = TRUE)

aveRDS(IPF_Mesenchymal, 
        file="../Desktop/Video_Tutorials/Data/GSE132771/IPF_Mesenchymal.RDS")

rm("features","IPF.anchors", "IPF.combined","IPF", "IPF.list","IPF_I", 
   "IPF_II", "IPF_III")

# 2. Convert to Seurat object to cell_data_set object ------
cds <- as.cell_data_set(IPF_Mesenchymal, group.by='ident') # SeuratWrappers
DefaultAssay(IPF_Mesenchymal)
DefaultAssay(IPF_Mesenchymal) <- "RNA"

# plot
colData(cds)
plot_cells(cds, color_cells_by = 'seurat_clusters',
           label_groups_by_cluster = FALSE,
           group_label_size = 5) +
  theme(legend.position = "right")

plot_cells(cds, color_cells_by = 'ident', 
           label_groups_by_cluster = FALSE,
           group_label_size = 5) +
  theme(legend.position = "right")

# Learn the trajectory graph
cds <- learn_graph(cds, use_partition = FALSE)

cds <- cluster_cells(cds, resolution = 1e-5)

plot_cells(cds, color_cells_by = 'ident', 
           label_groups_by_cluster = FALSE,
           group_label_size = 5) +
  theme(legend.position = "right")

# 3. trajectory analysis------
cds <- learn_graph(cds, use_partition = FALSE)

# plot
plot_cells(cds, color_cells_by = 'ident', 
           label_groups_by_cluster = FALSE,
           group_label_size = 5) +
  theme(legend.position = "right")

plot_cells(cds, genes=c("CD34", "PI16", "SCARA5", "MFAP5"))

# Create gene annotation file
fData(cds) # gene_annotation
rownames(fData(cds))[1:10]
fData(cds)$gene_short_name <- rownames(fData(cds))
fData(cds)

# Order the cells by pseudotime 
cds <- order_cells(cds)

#plot_cells in pseudotime
plot_cells(cds, color_cells_by = 'pseudotime')
```

# Lecture 4
```{r}
####### Constructing trajectories with accessibility data
# 1 Load the data
# 2 Preprocess the data
# 3 Reduce the dimensionality of the data
# 4 Cluster the cells
# 5 Learn the trajectory graph
# 6 Order the cells in pseudotime
# 7 Plot the results in pseudotime
# 8 Other analysis

library(cicero)
library(monocle3)
set.seed(2018)

#########################################################################
# 1. Loading 10X scATAC-seq data

# read in matrix data using the Matrix package
indata <- Matrix::readMM("../Desktop/Video_Tutorials/Data/filtered_peak_bc_matrix/matrix.mtx") 
indata

# binarize the matrix (One read per peak in a cell)
indata@x[indata@x > 0] <- 1
indata

# format cell info
cellinfo <- read.table("../Desktop/Video_Tutorials/Data/filtered_peak_bc_matrix/barcodes.tsv")
rownames(cellinfo) <- cellinfo$V1
names(cellinfo) <- "cells"

# format peak info
peakinfo <- read.table("../Desktop/Video_Tutorials/Data/filtered_peak_bc_matrix/peaks.bed")
names(peakinfo) <- c("chr", "bp1", "bp2")
peakinfo$site_name <- paste(peakinfo$chr, peakinfo$bp1, peakinfo$bp2, sep="_")
rownames(peakinfo) <- peakinfo$site_name

indata
rownames(indata) <- rownames(peakinfo)
colnames(indata) <- rownames(cellinfo)
indata

# make CDS
input_cds <- new_cell_data_set(indata,
                               cell_metadata = cellinfo,
                               gene_metadata = peakinfo)

input_cds <- monocle3::detect_genes(input_cds)

# Ensure there are no peaks included with zero reads
input_cds <- input_cds[Matrix::rowSums(exprs(input_cds)) != 0,] 

# 2 Preprocess the data
# 3 Reduce the dimensionality of the data
# 4 Cluster the cells
# 5 Learn the trajectory graph
# 6 Order the cells in pseudotime
# 7 Plot the results in pseudotime
# 8 Other analysis

#######################################################################################################
# download online kidney ATAC data 
# "http://staff.washington.edu/hpliner/data/kidney_data.txt.gz"

# 1. read in the data from a simple sparse matrix format
cicero_data <- read.table('../Desktop/Video_Tutorials/Data/kidney_data.txt.gz')

input_cds <- make_atac_cds(cicero_data) 

# 2. Preprocess and normalize the data using Latent Semantic Indexing (LSI)
input_cds <- estimate_size_factors(input_cds) 
           # normalization process to select an optimal baseline sample
input_cds <- preprocess_cds(input_cds, method = "LSI")

# 3. standard dimensionality reduction used using UMAP in Monocle 3
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', 
                              preprocess_method = "LSI")

# 4. Cluster the cells
input_cds <- cluster_cells(input_cds)
plot_cells(input_cds)

# 5. Learn the trajectory graph
input_cds <- learn_graph(input_cds)

# 6. cell ordering 
input_cds <- order_cells(input_cds)

# 7. Plot the results in pseudotime
plot_cells(input_cds, color_cells_by = "pseudotime")

# 8. other analysis: Visualizing accessibility across pseudotime
input_cds_lin <- input_cds[,is.finite(pseudotime(input_cds))]
plot_accessibility_in_pseudotime(input_cds_lin[c("chr1_3238849_3239700", 
                                                 "chr1_3406155_3407044", 
                                                 "chr1_3397204_3397842")])
```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```

