<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Part 1: Global Setup and Data Loading</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="mIBD_Master_Analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="mIBD_Master_Analysis_files/libs/quarto-html/quarto.js"></script>
<script src="mIBD_Master_Analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="mIBD_Master_Analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="mIBD_Master_Analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="mIBD_Master_Analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="mIBD_Master_Analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="mIBD_Master_Analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="mIBD_Master_Analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="mIBD_Master_Analysis_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Part 1: Global Setup and Data Loading</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="project-setup-and-variable-definitions" class="level2">
<h2 class="anchored" data-anchor-id="project-setup-and-variable-definitions">1.1: Project Setup and Variable Definitions</h2>
<p>This section loads all required libraries, sets up the R environment, and defines key variables based on the project’s data hierarchy.</p>
<section id="data-hierarchy" class="level3">
<h3 class="anchored" data-anchor-id="data-hierarchy">Data Hierarchy</h3>
<p>The data has the following structure: - <strong>Patient</strong>: The unique patient identifier (<code>emrid</code>). - <strong>Drug Treatment</strong>: A unique patient-drug treatment pair (<code>key</code>, derived from <code>emrid</code> and <code>drugno</code>). - <strong>Sample</strong>: The physical pathology slide (<code>patho_id</code>). - <strong>ROI (Region of Interest)</strong>: A region on the TMA slide (<code>key_tma</code>). - <strong>AOI (Area of Interest)</strong>: The final unit of observation from GeoMx, corresponding to a single cell/observation in the Seurat object. The identifier is the barcode/rowname (<code>SegmentDisplayName</code>).</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load myR package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>library(devtools)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Unload the package if it's already loaded to ensure we get the latest version from source</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="st">"myR"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> loadedNamespaces()) {</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  unloadNamespace(<span class="st">"myR"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the development version of the package from the specified path</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>devtools::load_all(<span class="st">"/data/kjc1/mylit/myR"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Core packages</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>library(Seurat)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>library(dplyr)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>library(tidyr)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>library(ggplot2)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>library(readxl)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>library(lme4)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>library(emmeans)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>library(purrr)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>library(ComplexHeatmap)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>library(pheatmap)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>library(grid)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>library(ggrepel)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>library(stringr)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>library(broom)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>library(car)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>library(Hmisc)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>library(circlize)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>library(patchwork)  <span class="co"># For side-by-side plots</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>library(moments)    <span class="co"># For skewness and kurtosis</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>library(goftest)     <span class="co"># For Anderson-Darling test</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>library(conflicted) <span class="co"># To manage namespace conflicts</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Set preferences for conflicting functions</span></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"filter"</span>, <span class="st">"dplyr"</span>)</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"lag"</span>, <span class="st">"dplyr"</span>)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"select"</span>, <span class="st">"dplyr"</span>)</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"summarise"</span>, <span class="st">"dplyr"</span>)</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"summarize"</span>, <span class="st">"dplyr"</span>)</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>conflict_prefer(<span class="st">"pheatmap"</span>, <span class="st">"pheatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define consistent column names based on the data hierarchy</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># - Patient (emrid): A unique patient identifier.</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - Patient-Drug Pair (key): A unique patient-drug treatment pair, derived from emrid and drugno.</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   A patient can have multiple entries if they try different drugs over time.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - Sample (patho_id): The physical pathology slide.</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - ROI (key_tma): A Region of Interest on the TMA slide. This is the key used for pre/post pairing.</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># - AOI (SegmentDisplayName): The final unit of observation (Area of Interest), corresponding to a "cell" in Seurat.</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>VAR_PATIENT_ID <span class="op">&lt;-</span> <span class="st">"emrid"</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>VAR_SAMPLE_ID  <span class="op">&lt;-</span> <span class="st">"patho_id"</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>VAR_ROI_ID     <span class="op">&lt;-</span> <span class="st">"key_tma"</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>VAR_AOI_ID     <span class="op">&lt;-</span> <span class="st">"SegmentDisplayName"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>VAR_DRUG       <span class="op">&lt;-</span> <span class="st">"drug"</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>VAR_RESPONSE_CRITERION <span class="op">&lt;-</span> <span class="st">"Endo remission"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>PROJECT_ROOT <span class="op">&lt;-</span> <span class="st">"/data/kjc1/projects/#124. mIBD"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="load-raw-data" class="level2">
<h2 class="anchored" data-anchor-id="load-raw-data">1.2: Load Raw Data</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load count matrix</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>mibd <span class="op">&lt;-</span> read_xlsx(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"Initial Dataset_merged_250729.xlsx"</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  sheet <span class="op">=</span> <span class="st">"BioProbeCountMatrix"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>) <span class="op">%&gt;%</span> <span class="im">as</span>.data.frame()</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Load segment properties (raw metadata)</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>mibd_meta_raw <span class="op">&lt;-</span> read_xlsx(</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"Initial Dataset_merged_250729.xlsx"</span>),</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  sheet <span class="op">=</span> <span class="st">"SegmentProperties"</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>) <span class="op">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  .[<span class="op">!</span>names(.) <span class="op">%</span><span class="kw">in</span><span class="op">%</span> c(<span class="st">"drug"</span>, <span class="st">"treatment"</span>)] <span class="op">%&gt;%</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="im">as</span>.data.frame()</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Load clinical metadata</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>mibd_meta_clinical <span class="op">&lt;-</span> read_xlsx(</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"1. biologics naive 이름_병록번호 제외_merged2_250728.xlsx"</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>) <span class="op">%&gt;%</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>  .[<span class="op">!</span>names(.) <span class="op">%</span><span class="kw">in</span><span class="op">%</span> c(<span class="st">"treatment"</span>, <span class="st">"slide"</span>, <span class="st">"TMA 순번"</span>)] <span class="op">%&gt;%</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>  <span class="im">as</span>.data.frame()</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create long format for pre/post matching</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="co"># 'key_pre' and 'key_post' are the Ground Truth for pairing, defined in the clinical data.</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co"># We pivot this data to create a long-format table.</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="co"># CRITICAL: We also explicitly create the 'treatment' column here.</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>mibd_meta_clinical_long <span class="op">&lt;-</span> mibd_meta_clinical <span class="op">%&gt;%</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  pivot_longer(</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    cols <span class="op">=</span> c(key_pre, key_post),</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    names_to <span class="op">=</span> <span class="st">"treatment"</span>, <span class="co"># This will contain "key_pre" or "key_post"</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    values_to <span class="op">=</span> <span class="st">"key_tma"</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>  mutate(</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Clean up the 'treatment' column to be just 'pre' or 'post'</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>    treatment <span class="op">=</span> if_else(treatment <span class="op">==</span> <span class="st">"key_pre"</span>, <span class="st">"pre"</span>, <span class="st">"post"</span>)</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Join metadata</span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>mibd_meta <span class="op">&lt;-</span> mibd_meta_raw <span class="op">%&gt;%</span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>  mutate(ROILabel <span class="op">=</span> <span class="im">as</span>.numeric(ROILabel)) <span class="op">%&gt;%</span></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>  left_join(mibd_meta_clinical_long, by <span class="op">=</span> c(<span class="st">"key_tma"</span>))</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Final metadata with response information</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>mibd_meta_final <span class="op">&lt;-</span> mibd_meta <span class="op">%&gt;%</span></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>  drop_na(patho_id, `Endo remission`) <span class="op">%&gt;%</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>  mutate(response <span class="op">=</span> `Endo remission`)</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare data matrices</span></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a>raw_data <span class="op">&lt;-</span> mibd</span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>rownames(raw_data) <span class="op">&lt;-</span> raw_data$ProbeDisplayName</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>SP <span class="op">&lt;-</span> mibd_meta_final</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>rownames(SP) <span class="op">&lt;-</span> SP$SegmentDisplayName</span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Data loaded successfully:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Gene probes:"</span>, nrow(raw_data), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Samples (ROIs):"</span>, nrow(SP), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Patients:"</span>, length(unique(SP$emrid)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="gene-and-aoi-filtering" class="level2">
<h2 class="anchored" data-anchor-id="gene-and-aoi-filtering">1.2 Gene and AOI Filtering</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define column choice</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>choice_of_column <span class="op">&lt;-</span> <span class="st">"TargetName"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Select relevant columns</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>data_selected <span class="op">&lt;-</span> raw_data <span class="op">%&gt;%</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  dplyr::select(</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    `ProbeDisplayName`,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    `TargetName`,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    `HUGOSymbol`,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    matches(<span class="st">"</span><span class="ch">\\</span><span class="st">|"</span>)  <span class="co"># ROI columns</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify numeric columns (AOI counts)</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>numeric_cols <span class="op">&lt;-</span> names(data_selected)[sapply(data_selected, <span class="kw">is</span>.numeric)]</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>meta_cols <span class="op">&lt;-</span> c(<span class="st">"ProbeDisplayName"</span>, <span class="st">"TargetName"</span>, <span class="st">"HUGOSymbol"</span>)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>aoi_cols <span class="op">&lt;-</span> setdiff(numeric_cols, meta_cols)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>aoi_cols <span class="op">&lt;-</span> intersect(aoi_cols, rownames(SP))</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate AOI totals for QC</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>aoi_sum <span class="op">&lt;-</span> colSums(data_selected[aoi_cols], na.rm <span class="op">=</span> TRUE)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter AOIs by IQR criterion</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>q1 <span class="op">&lt;-</span> quantile(aoi_sum, <span class="fl">0.25</span>)</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>q3 <span class="op">&lt;-</span> quantile(aoi_sum, <span class="fl">0.75</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>iqr_val <span class="op">&lt;-</span> q3 <span class="op">-</span> q1</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>low_cutoff <span class="op">&lt;-</span> q1 <span class="op">-</span> <span class="fl">1.5</span> <span class="op">*</span> iqr_val</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>remove_aoi <span class="op">&lt;-</span> names(aoi_sum[aoi_sum <span class="op">&lt;</span> low_cutoff])</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"AOI filtering:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Total AOIs:"</span>, length(aoi_cols), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- AOIs removed:"</span>, length(remove_aoi), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- AOIs kept:"</span>, length(aoi_cols) <span class="op">-</span> length(remove_aoi), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply AOI filtering</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>data_filtered_aoi <span class="op">&lt;-</span> data_selected <span class="op">%&gt;%</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>  select(<span class="op">-</span>all_of(remove_aoi)) <span class="op">%&gt;%</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>  select(`ProbeDisplayName`, `TargetName`, `HUGOSymbol`, all_of(aoi_cols))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="negative-probe-processing" class="level2">
<h2 class="anchored" data-anchor-id="negative-probe-processing">1.3 Negative Probe Processing</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract negative probes for background estimation</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>neg_data <span class="op">&lt;-</span> data_filtered_aoi <span class="op">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(grepl(<span class="st">"NegProbe-WTX"</span>, ProbeDisplayName))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>neg_aoi <span class="op">&lt;-</span> neg_data[intersect(aoi_cols, names(neg_data))]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>neg_geo_mean <span class="op">&lt;-</span> exp(rowMeans(log(neg_aoi <span class="op">+</span> <span class="dv">1</span>), na.rm <span class="op">=</span> TRUE))  <span class="co"># +1 to avoid log(0)</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>neg_overall_geo_mean <span class="op">&lt;-</span> exp(mean(log(neg_geo_mean), na.rm <span class="op">=</span> TRUE))  <span class="co"># Geometric mean</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>neg_log_sd <span class="op">&lt;-</span> sd(log(neg_geo_mean), na.rm <span class="op">=</span> TRUE)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>LOQ <span class="op">&lt;-</span> exp(log(neg_overall_geo_mean) <span class="op">+</span> <span class="dv">2</span> <span class="op">*</span> neg_log_sd)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Negative probe statistics:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Geometric mean:"</span>, <span class="bu">round</span>(neg_overall_geo_mean, <span class="dv">2</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- LOQ (Limit of Quantification):"</span>, <span class="bu">round</span>(LOQ, <span class="dv">2</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="biological-probe-processing-and-qc" class="level2">
<h2 class="anchored" data-anchor-id="biological-probe-processing-and-qc">1.4 Biological Probe Processing and QC</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter to biological probes only</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>bio_data <span class="op">&lt;-</span> data_filtered_aoi <span class="op">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(<span class="op">!</span>grepl(<span class="st">"NegProbe-WTX"</span>, ProbeDisplayName))</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>bio_aoi <span class="op">&lt;-</span> bio_data[intersect(aoi_cols, names(bio_data))]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>probe_geo_mean <span class="op">&lt;-</span> exp(rowMeans(log(bio_aoi <span class="op">+</span> <span class="dv">1</span>), na.rm <span class="op">=</span> TRUE))  <span class="co"># +1 to avoid log(0)</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate ratio to target for QC</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>overall_geo_mean <span class="op">&lt;-</span> exp(mean(log(probe_geo_mean), na.rm <span class="op">=</span> TRUE))  <span class="co"># Geometric mean</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>ratio_to_target <span class="op">&lt;-</span> probe_geo_mean <span class="op">/</span> overall_geo_mean</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter probes below 10% of geomean</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>cutoff <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>keep_probes <span class="op">&lt;-</span> ratio_to_target <span class="op">&gt;</span> cutoff</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>bio_data_qc <span class="op">&lt;-</span> bio_data[keep_probes, ]</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Probe QC:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Starting probes:"</span>, nrow(bio_data), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Probes removed:"</span>, <span class="bu">sum</span>(<span class="op">!</span>keep_probes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Probes kept:"</span>, <span class="bu">sum</span>(keep_probes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="q3-normalization" class="level2">
<h2 class="anchored" data-anchor-id="q3-normalization">1.5 Q3 Normalization</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare data for Q3 normalization</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># First, select only numeric AOI columns from bio_data_qc</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>aoi_cols_clean <span class="op">&lt;-</span> intersect(aoi_cols, names(bio_data_qc))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>expr_matrix <span class="op">&lt;-</span> bio_data_qc <span class="op">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  select(all_of(aoi_cols_clean)) <span class="op">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="im">as</span>.matrix()</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Set rownames to gene names</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>rownames(expr_matrix) <span class="op">&lt;-</span> bio_data_qc$TargetName</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create gene info</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>gene_info <span class="op">&lt;-</span> bio_data_qc <span class="op">%&gt;%</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  select(ProbeDisplayName, TargetName, HUGOSymbol) <span class="op">%&gt;%</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  mutate(gene <span class="op">=</span> TargetName)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Verify matrix dimensions</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Preparing expression matrix:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Genes (rows):"</span>, nrow(expr_matrix), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Samples (columns):"</span>, ncol(expr_matrix), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Sample names:"</span>, head(colnames(expr_matrix)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply Q3 normalization directly to the matrix</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="co"># The q3_normalize function from GeoMx.R takes a matrix and returns a list</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>geomx_norm <span class="op">&lt;-</span> myR::q3_normalize(expr_matrix, scaling_factor <span class="op">=</span> <span class="dv">1000</span>)</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract normalized counts</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>norm_expr <span class="op">&lt;-</span> geomx_norm$normalized  <span class="co"># Linear scale normalized</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>norm_expr_log <span class="op">&lt;-</span> geomx_norm$normalized_log  <span class="co"># Log2 transformed</span></span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a><span class="co"># LOG TRANSFORMATION LOGIC:</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Q3 normalization scales counts to have similar Q3 (75th percentile) values across samples.</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Log transformation (log2) is then applied because:</span></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Count data is typically right-skewed - log makes distribution more normal</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Compresses large values and expands small values - reduces impact of outliers</span></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Makes variance more stable across expression levels (variance stabilization)</span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Standard practice in RNA-seq/spatial transcriptomics for downstream analysis</span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. Q3 normalization brings samples to similar scales, then log2 makes them </span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a><span class="co">#    comparable and suitable for statistical tests (t-tests, linear models, etc.)</span></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. Log transform also helps with fold-change interpretation (log2 FC = doubling)</span></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Normalization complete:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Input matrix:"</span>, dim(expr_matrix), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Normalized matrix:"</span>, dim(norm_expr), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Q3 factors applied:"</span>, length(geomx_norm$q3_factors), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Note: norm_expr = linear scale Q3-normalized counts</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"      norm_expr_log = log2(norm_expr + 1) for downstream analysis</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="normalization-verification" class="level2">
<h2 class="anchored" data-anchor-id="normalization-verification">1.6 Normalization Verification</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure SP row names match column names of expression matrix</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="op">!</span>identical(rownames(SP), colnames(expr_matrix))) {</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  rownames(SP) <span class="op">&lt;-</span> colnames(expr_matrix)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Seurat object for visualization (Seurat v5 compatible)</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>seurat_raw <span class="op">&lt;-</span> CreateSeuratObject(</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  counts <span class="op">=</span> expr_matrix,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  project <span class="op">=</span> <span class="st">"mIBD"</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  meta.data <span class="op">=</span> SP</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create normalized Seurat object</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>seurat_norm <span class="op">&lt;-</span> CreateSeuratObject(</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  counts <span class="op">=</span> norm_expr,  <span class="co"># Use normalized counts (linear scale)</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  project <span class="op">=</span> <span class="st">"mIBD"</span>,</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  meta.data <span class="op">=</span> SP</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Set log-normalized data as a layer (Seurat v5 uses layers, not slots)</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>seurat_norm <span class="op">&lt;-</span> SetAssayData(</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  seurat_norm,</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>  layer <span class="op">=</span> <span class="st">"data"</span>,</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>  new.data <span class="op">=</span> <span class="im">as</span>.matrix(norm_expr_log),</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate QC metrics using Layers() for Seurat v5</span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>seurat_raw$total_counts <span class="op">&lt;-</span> colSums(LayerData(seurat_raw, layer <span class="op">=</span> <span class="st">"counts"</span>))</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>seurat_norm$total_counts <span class="op">&lt;-</span> colSums(LayerData(seurat_norm, layer <span class="op">=</span> <span class="st">"counts"</span>))</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize QC</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>p1 <span class="op">&lt;-</span> myR::mybar(seurat_raw<span class="op">@</span>meta.data, column <span class="op">=</span> total_counts, </span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>            main <span class="op">=</span> <span class="st">"Total Counts Before Normalization"</span>)</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>p2 <span class="op">&lt;-</span> myR::mybar(seurat_norm<span class="op">@</span>meta.data, column <span class="op">=</span> total_counts,</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>            main <span class="op">=</span> <span class="st">"Total Counts After Q3 Normalization"</span>)</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Show plots side by side</span></span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>library(patchwork)</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>combined_plot <span class="op">&lt;-</span> p1 <span class="op">|</span> p2</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(combined_plot)</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Save plots to project folder</span></span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="bu">dir</span>.exists(<span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>))) {</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>  <span class="bu">dir</span>.create(<span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>), recursive <span class="op">=</span> TRUE)</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Save individual plots</span></span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>ggsave(<span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>, <span class="st">"raw_counts_histogram.png"</span>), </span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>       p1, width <span class="op">=</span> <span class="dv">8</span>, height <span class="op">=</span> <span class="dv">6</span>, dpi <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>ggsave(<span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>, <span class="st">"normalized_counts_histogram.png"</span>), </span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>       p2, width <span class="op">=</span> <span class="dv">8</span>, height <span class="op">=</span> <span class="dv">6</span>, dpi <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>ggsave(<span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>, <span class="st">"counts_comparison.png"</span>), </span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>       combined_plot, width <span class="op">=</span> <span class="dv">12</span>, height <span class="op">=</span> <span class="dv">6</span>, dpi <span class="op">=</span> <span class="dv">300</span>)</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Plots saved to:"</span>, <span class="bu">file</span>.path(PROJECT_ROOT, <span class="st">"plots"</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Distribution testing</span></span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>test_distribution <span class="op">&lt;-</span> function(data, name <span class="op">=</span> <span class="st">"Data"</span>) {</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">=== Distribution Testing for"</span>, name, <span class="st">"===</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove zeros and NAs for testing</span></span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>  test_data <span class="op">&lt;-</span> data[data <span class="op">&gt;</span> <span class="dv">0</span> <span class="op">&amp;</span> <span class="op">!</span><span class="kw">is</span>.na(data)]</span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (length(test_data) <span class="op">&lt;</span> <span class="dv">3</span>) {</span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"Insufficient data for testing</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(NULL)</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Shapiro-Wilk test for normality</span></span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a>  shapiro_test <span class="op">&lt;-</span> shapiro.test(test_data)</span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"Shapiro-Wilk normality test:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  p-value:"</span>, <span class="bu">round</span>(shapiro_test$p.value, <span class="dv">6</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Normal:"</span>, shapiro_test$p.value <span class="op">&gt;</span> <span class="fl">0.05</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Kolmogorov-Smirnov test for log-normal</span></span>
<span id="cb8-79"><a href="#cb8-79" aria-hidden="true" tabindex="-1"></a>  log_data <span class="op">&lt;-</span> log(test_data)</span>
<span id="cb8-80"><a href="#cb8-80" aria-hidden="true" tabindex="-1"></a>  ks_log <span class="op">&lt;-</span> ks.test(log_data, <span class="st">"pnorm"</span>, mean(log_data), sd(log_data))</span>
<span id="cb8-81"><a href="#cb8-81" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Kolmogorov-Smirnov test for log-normality:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-82"><a href="#cb8-82" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  p-value:"</span>, <span class="bu">round</span>(ks_log$p.value, <span class="dv">6</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-83"><a href="#cb8-83" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Log-normal:"</span>, ks_log$p.value <span class="op">&gt;</span> <span class="fl">0.05</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-84"><a href="#cb8-84" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-85"><a href="#cb8-85" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Anderson-Darling test for exponential (if available)</span></span>
<span id="cb8-86"><a href="#cb8-86" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (requireNamespace(<span class="st">"goftest"</span>, quietly <span class="op">=</span> TRUE)) {</span>
<span id="cb8-87"><a href="#cb8-87" aria-hidden="true" tabindex="-1"></a>    exp_test <span class="op">&lt;-</span> goftest::ad.test(test_data, <span class="st">"pexp"</span>)</span>
<span id="cb8-88"><a href="#cb8-88" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Anderson-Darling test for exponential:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-89"><a href="#cb8-89" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  p-value:"</span>, <span class="bu">round</span>(exp_test$p.value, <span class="dv">6</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-90"><a href="#cb8-90" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  Exponential:"</span>, exp_test$p.value <span class="op">&gt;</span> <span class="fl">0.05</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-91"><a href="#cb8-91" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-92"><a href="#cb8-92" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-93"><a href="#cb8-93" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Basic statistics</span></span>
<span id="cb8-94"><a href="#cb8-94" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Basic statistics:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-95"><a href="#cb8-95" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Mean:"</span>, <span class="bu">round</span>(mean(test_data), <span class="dv">2</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-96"><a href="#cb8-96" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Median:"</span>, <span class="bu">round</span>(median(test_data), <span class="dv">2</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-97"><a href="#cb8-97" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  SD:"</span>, <span class="bu">round</span>(sd(test_data), <span class="dv">2</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-98"><a href="#cb8-98" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-99"><a href="#cb8-99" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate skewness and kurtosis manually if moments package not available</span></span>
<span id="cb8-100"><a href="#cb8-100" aria-hidden="true" tabindex="-1"></a>  n <span class="op">&lt;-</span> length(test_data)</span>
<span id="cb8-101"><a href="#cb8-101" aria-hidden="true" tabindex="-1"></a>  mean_val <span class="op">&lt;-</span> mean(test_data)</span>
<span id="cb8-102"><a href="#cb8-102" aria-hidden="true" tabindex="-1"></a>  sd_val <span class="op">&lt;-</span> sd(test_data)</span>
<span id="cb8-103"><a href="#cb8-103" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-104"><a href="#cb8-104" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Skewness calculation</span></span>
<span id="cb8-105"><a href="#cb8-105" aria-hidden="true" tabindex="-1"></a>  skewness <span class="op">&lt;-</span> <span class="bu">sum</span>((test_data <span class="op">-</span> mean_val)<span class="op">^</span><span class="dv">3</span>) <span class="op">/</span> (n <span class="op">*</span> sd_val<span class="op">^</span><span class="dv">3</span>)</span>
<span id="cb8-106"><a href="#cb8-106" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Skewness:"</span>, <span class="bu">round</span>(skewness, <span class="dv">3</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-107"><a href="#cb8-107" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-108"><a href="#cb8-108" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Kurtosis calculation</span></span>
<span id="cb8-109"><a href="#cb8-109" aria-hidden="true" tabindex="-1"></a>  kurtosis <span class="op">&lt;-</span> <span class="bu">sum</span>((test_data <span class="op">-</span> mean_val)<span class="op">^</span><span class="dv">4</span>) <span class="op">/</span> (n <span class="op">*</span> sd_val<span class="op">^</span><span class="dv">4</span>) <span class="op">-</span> <span class="dv">3</span></span>
<span id="cb8-110"><a href="#cb8-110" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"  Kurtosis:"</span>, <span class="bu">round</span>(kurtosis, <span class="dv">3</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-111"><a href="#cb8-111" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-112"><a href="#cb8-112" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Determine best fit</span></span>
<span id="cb8-113"><a href="#cb8-113" aria-hidden="true" tabindex="-1"></a>  best_fit <span class="op">&lt;-</span> <span class="st">"Unknown"</span></span>
<span id="cb8-114"><a href="#cb8-114" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (shapiro_test$p.value <span class="op">&gt;</span> <span class="fl">0.05</span>) {</span>
<span id="cb8-115"><a href="#cb8-115" aria-hidden="true" tabindex="-1"></a>    best_fit <span class="op">&lt;-</span> <span class="st">"Normal"</span></span>
<span id="cb8-116"><a href="#cb8-116" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (ks_log$p.value <span class="op">&gt;</span> <span class="fl">0.05</span>) {</span>
<span id="cb8-117"><a href="#cb8-117" aria-hidden="true" tabindex="-1"></a>    best_fit <span class="op">&lt;-</span> <span class="st">"Log-normal"</span></span>
<span id="cb8-118"><a href="#cb8-118" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb8-119"><a href="#cb8-119" aria-hidden="true" tabindex="-1"></a>    best_fit <span class="op">&lt;-</span> <span class="st">"Non-normal (possibly exponential or other)"</span></span>
<span id="cb8-120"><a href="#cb8-120" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-121"><a href="#cb8-121" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-122"><a href="#cb8-122" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Best fit:"</span>, best_fit, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb8-123"><a href="#cb8-123" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-124"><a href="#cb8-124" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(<span class="bu">list</span>(</span>
<span id="cb8-125"><a href="#cb8-125" aria-hidden="true" tabindex="-1"></a>    shapiro <span class="op">=</span> shapiro_test,</span>
<span id="cb8-126"><a href="#cb8-126" aria-hidden="true" tabindex="-1"></a>    ks_log <span class="op">=</span> ks_log,</span>
<span id="cb8-127"><a href="#cb8-127" aria-hidden="true" tabindex="-1"></a>    best_fit <span class="op">=</span> best_fit,</span>
<span id="cb8-128"><a href="#cb8-128" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> <span class="bu">list</span>(</span>
<span id="cb8-129"><a href="#cb8-129" aria-hidden="true" tabindex="-1"></a>      mean <span class="op">=</span> mean(test_data),</span>
<span id="cb8-130"><a href="#cb8-130" aria-hidden="true" tabindex="-1"></a>      median <span class="op">=</span> median(test_data),</span>
<span id="cb8-131"><a href="#cb8-131" aria-hidden="true" tabindex="-1"></a>      sd <span class="op">=</span> sd(test_data),</span>
<span id="cb8-132"><a href="#cb8-132" aria-hidden="true" tabindex="-1"></a>      skewness <span class="op">=</span> skewness,</span>
<span id="cb8-133"><a href="#cb8-133" aria-hidden="true" tabindex="-1"></a>      kurtosis <span class="op">=</span> kurtosis</span>
<span id="cb8-134"><a href="#cb8-134" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-135"><a href="#cb8-135" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb8-136"><a href="#cb8-136" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-137"><a href="#cb8-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-138"><a href="#cb8-138" aria-hidden="true" tabindex="-1"></a><span class="co"># Test distributions</span></span>
<span id="cb8-139"><a href="#cb8-139" aria-hidden="true" tabindex="-1"></a>raw_dist <span class="op">&lt;-</span> test_distribution(seurat_raw$total_counts, <span class="st">"Raw Counts"</span>)</span>
<span id="cb8-140"><a href="#cb8-140" aria-hidden="true" tabindex="-1"></a>norm_dist <span class="op">&lt;-</span> test_distribution(seurat_norm$total_counts, <span class="st">"Normalized Counts"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="part-2-metadata-processing-and-analysis" class="level1">
<h1>Part 2: Metadata Processing and Analysis</h1>
<section id="metadata-pruning-and-filtering" class="level2">
<h2 class="anchored" data-anchor-id="metadata-pruning-and-filtering">2.1 Metadata Pruning and Filtering</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create patient-drug response summary</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>pd <span class="op">&lt;-</span> mibd_meta_final <span class="op">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(<span class="op">!</span><span class="kw">is</span>.na(response)) <span class="op">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  mutate(response <span class="op">=</span> <span class="im">as</span>.integer(response)) <span class="op">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  group_by(emrid, drug) <span class="op">%&gt;%</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  summarise(</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    vals <span class="op">=</span> unique(response),</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    patient_response <span class="op">=</span> <span class="cf">if</span> (length(vals) <span class="op">==</span> <span class="dv">1</span>) vals <span class="cf">else</span> NA_integer_,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    .groups <span class="op">=</span> <span class="st">"drop"</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  select(emrid, drug, patient_response)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Count responders by drug</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>counts <span class="op">&lt;-</span> pd <span class="op">%&gt;%</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(<span class="op">!</span><span class="kw">is</span>.na(patient_response)) <span class="op">%&gt;%</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  count(drug, patient_response, name <span class="op">=</span> <span class="st">"n_patients"</span>) <span class="op">%&gt;%</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  mutate(patient_response <span class="op">=</span> <span class="im">as</span>.character(patient_response)) <span class="op">%&gt;%</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  pivot_wider(</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>    names_from <span class="op">=</span> patient_response,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    values_from <span class="op">=</span> n_patients,</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    values_fill <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  rename(Responder <span class="op">=</span> `<span class="dv">1</span>`, NonResponder <span class="op">=</span> `<span class="dv">0</span>`) <span class="op">%&gt;%</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>  mutate(</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    Total <span class="op">=</span> Responder <span class="op">+</span> NonResponder,</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    RespRate <span class="op">=</span> ifelse(Total <span class="op">&gt;</span> <span class="dv">0</span>, Responder <span class="op">/</span> Total, NA_real_)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  ) <span class="op">%&gt;%</span></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>  arrange(drug)</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="metadata-multicollinearity-analysis" class="level2">
<h2 class="anchored" data-anchor-id="metadata-multicollinearity-analysis">2.2 Metadata Multicollinearity Analysis</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare metadata for correlation analysis</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>meta_vars <span class="op">&lt;-</span> SP <span class="op">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  select_if(<span class="kw">is</span>.numeric) <span class="op">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  select(<span class="op">-</span>contains(<span class="st">"segment"</span>))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate correlation matrix</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>cor_matrix <span class="op">&lt;-</span> cor(meta_vars, use <span class="op">=</span> <span class="st">"pairwise.complete.obs"</span>, method <span class="op">=</span> <span class="st">"pearson"</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize multicollinearity</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>p <span class="op">&lt;-</span> pheatmap(</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  cor_matrix,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  clustering_distance_rows <span class="op">=</span> <span class="st">"correlation"</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  clustering_distance_cols <span class="op">=</span> <span class="st">"correlation"</span>,</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  main <span class="op">=</span> <span class="st">"Metadata Variable Correlations"</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="metadata-merging" class="level2">
<h2 class="anchored" data-anchor-id="metadata-merging">2.3 Metadata Merging</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create comprehensive metadata with response info</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Store original treatment column before modification</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>SP_interim <span class="op">&lt;-</span> SP <span class="op">%&gt;%</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  mutate(treatment_original <span class="op">=</span> treatment) <span class="op">%&gt;%</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  left_join(pd <span class="op">%&gt;%</span> select(emrid, drug, patient_response),</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>            by <span class="op">=</span> c(<span class="st">"emrid"</span>, <span class="st">"drug"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  mutate(</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    response_status <span class="op">=</span> case_when(</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>      patient_response <span class="op">==</span> <span class="dv">1</span> <span class="op">~</span> <span class="st">"Responder"</span>,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>      patient_response <span class="op">==</span> <span class="dv">0</span> <span class="op">~</span> <span class="st">"NonResponder"</span>,</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>      TRUE <span class="op">~</span> NA_character_</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    treatment_combined <span class="op">=</span> ifelse(<span class="op">!</span><span class="kw">is</span>.na(drug) <span class="op">&amp;</span> <span class="op">!</span><span class="kw">is</span>.na(treatment_original),</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>                               paste0(drug, <span class="st">"_"</span>, treatment_original),</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>                               treatment_original)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="co"># DEBUGGING ROW NAMES</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"=== DEBUGGING METADATA ROWNAMES ===</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Original SP rownames (first 5):"</span>, head(rownames(SP)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Seurat cell names (first 5):"</span>, head(colnames(seurat_norm)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if row names were lost during manipulation</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"SP_interim rownames after join:"</span>, head(rownames(SP_interim)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(identical(<span class="im">as</span>.character(<span class="dv">1</span>:nrow(SP_interim)), rownames(SP_interim))) {</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"WARNING: Row names were reset to default integers!</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a><span class="co"># The row names for the metadata MUST match the cell names in the Seurat object.</span></span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a><span class="co"># The original SP dataframe had correct rownames. We need to restore them.</span></span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>SP_final <span class="op">&lt;-</span> SP_interim</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>rownames(SP_final) <span class="op">&lt;-</span> rownames(SP)</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Restored rownames to SP_final from original SP</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"SP_final rownames (first 5):"</span>, head(rownames(SP_final)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Is identical now?"</span>, identical(rownames(SP_final), colnames(seurat_norm)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Now that rownames are correct, update Seurat metadata</span></span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>seurat_norm<span class="op">@</span>meta.data <span class="op">&lt;-</span> SP_final</span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Final check</span></span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Final check: seurat_norm metadata rownames (first 5):"</span>, head(rownames(seurat_norm<span class="op">@</span>meta.data)), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Is final metadata identical to SP_final?"</span>, identical(seurat_norm<span class="op">@</span>meta.data, SP_final), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Final metadata summary:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Total samples:"</span>, nrow(SP_final), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Responders:"</span>, <span class="bu">sum</span>(SP_final$patient_response <span class="op">==</span> <span class="dv">1</span>, na.rm <span class="op">=</span> TRUE), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Non-responders:"</span>, <span class="bu">sum</span>(SP_final$patient_response <span class="op">==</span> <span class="dv">0</span>, na.rm <span class="op">=</span> TRUE), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Drugs:"</span>, paste(unique(SP_final$drug), collapse <span class="op">=</span> <span class="st">", "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="part-3-signature-scoring" class="level1">
<h1>Part 3: Signature Scoring</h1>
<section id="tbd-known-issues-and-workarounds" class="level2">
<h2 class="anchored" data-anchor-id="tbd-known-issues-and-workarounds">3.1: (TBD) Known Issues and Workarounds</h2>
<p>This section documents known issues encountered during the signature scoring process and the workarounds implemented in the code.</p>
<section id="addmodulescore-error-subscript-out-of-bounds" class="level3">
<h3 class="anchored" data-anchor-id="addmodulescore-error-subscript-out-of-bounds"><code>AddModuleScore</code> Error: <code>subscript out of bounds</code></h3>
<ul>
<li><strong>Problem</strong>: When calculating module scores (e.g., Inflammatory Score), Seurat’s <code>AddModuleScore</code> function sometimes fails with a <code>subscript out of bounds</code> error. This is a known, intermittent issue in Seurat, often related to the internal binning of genes by expression level to select control genes.</li>
<li><strong>Workaround</strong>: A <code>tryCatch</code> block is implemented. If <code>AddModuleScore</code> fails, the code falls back to a more direct method: calculating the mean expression of all available genes in the signature. While this doesn’t use Seurat’s control gene correction, it provides a robust and reliable score, preventing the analysis from halting.</li>
</ul>
</section>
<section id="progeny-warning-layer-counts-isnt-present" class="level3">
<h3 class="anchored" data-anchor-id="progeny-warning-layer-counts-isnt-present"><code>progeny</code> Warning: <code>Layer "counts" isn't present</code></h3>
<ul>
<li><strong>Problem</strong>: When running <code>progeny</code> to calculate pathway scores, a warning <code>Layer "counts" isn't present in the assay object</code> is generated. This is likely due to a minor incompatibility between the <code>progeny</code> package and the Seurat v5 object structure, which uses “layers” (e.g., <code>counts</code>, <code>data</code>) for storing different data matrices. The <code>progeny</code> function appears to specifically look for a <code>counts</code> layer which might not be the active or intended layer for its calculation.</li>
<li><strong>Workaround</strong>: The <code>myR::add_progeny_scores</code> wrapper function successfully proceeds by using the scaled data from the <code>progeny</code> assay it generates. The warning can be safely ignored as the final scores are correctly calculated and added to the object metadata.</li>
</ul>
<hr>
</section>
</section>
<section id="calculate-inflammatory-score-is" class="level2">
<h2 class="anchored" data-anchor-id="calculate-inflammatory-score-is">3.2 Calculate Inflammatory Score (IS)</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define inflammatory signature genes</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>inflammatory_genes <span class="op">&lt;-</span> c(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"IL1B"</span>, <span class="st">"TNF"</span>, <span class="st">"IL6"</span>, <span class="st">"CXCL8"</span>, <span class="st">"CXCL1"</span>, <span class="st">"CXCL2"</span>, <span class="st">"CXCL3"</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"CCL2"</span>, <span class="st">"CCL3"</span>, <span class="st">"CCL4"</span>, <span class="st">"CCL5"</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"S100A8"</span>, <span class="st">"S100A9"</span>, <span class="st">"S100A12"</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"CD68"</span>, <span class="st">"CD163"</span>, <span class="st">"ARG1"</span>, <span class="st">"NOS2"</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Add module score for inflammatory genes</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Checking gene availability:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Total genes in Seurat object:"</span>, nrow(seurat_norm), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- First 10 gene names:"</span>, head(rownames(seurat_norm), <span class="dv">10</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which inflammatory genes are available</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>available_genes <span class="op">&lt;-</span> inflammatory_genes[inflammatory_genes <span class="op">%</span><span class="kw">in</span><span class="op">%</span> rownames(seurat_norm)]</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>missing_genes <span class="op">&lt;-</span> setdiff(inflammatory_genes, available_genes)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Requested inflammatory genes:"</span>, length(inflammatory_genes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Available genes:"</span>, length(available_genes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Missing genes:"</span>, paste(missing_genes, collapse <span class="op">=</span> <span class="st">", "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (length(available_genes) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"- Available genes:"</span>, paste(available_genes, collapse <span class="op">=</span> <span class="st">", "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create gene list for AddModuleScore</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>  gene_list <span class="op">&lt;-</span> <span class="bu">list</span>(Inflammatory <span class="op">=</span> available_genes)</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Try AddModuleScore with error handling</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  tryCatch({</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>    seurat_norm <span class="op">&lt;-</span> AddModuleScore(</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>      seurat_norm,</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>      features <span class="op">=</span> gene_list,</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>      name <span class="op">=</span> <span class="st">"IS_"</span>,</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>      assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rename the module score column</span></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>    seurat_norm$IS <span class="op">&lt;-</span> seurat_norm$IS_1</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"Inflammatory Score added successfully:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"- Signature genes used:"</span>, length(available_genes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"- IS range:"</span>, paste(<span class="bu">round</span>(<span class="bu">range</span>(seurat_norm$IS, na.rm <span class="op">=</span> TRUE), <span class="dv">3</span>), collapse <span class="op">=</span> <span class="st">" to "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>  }, error <span class="op">=</span> function(e) {</span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"Error in AddModuleScore:"</span>, e$message, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"Trying alternative approach...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Alternative: Calculate IS manually</span></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>    expr_data <span class="op">&lt;-</span> LayerData(seurat_norm, layer <span class="op">=</span> <span class="st">"data"</span>)</span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>    available_expr <span class="op">&lt;-</span> expr_data[available_genes, ]</span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>    seurat_norm$IS <span class="op">&lt;-</span> colMeans(available_expr, na.rm <span class="op">=</span> TRUE)</span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"Inflammatory Score calculated manually:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"- Signature genes used:"</span>, length(available_genes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"- IS range:"</span>, paste(<span class="bu">round</span>(<span class="bu">range</span>(seurat_norm$IS, na.rm <span class="op">=</span> TRUE), <span class="dv">3</span>), collapse <span class="op">=</span> <span class="st">" to "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>  warning(<span class="st">"No inflammatory genes found in the dataset!"</span>)</span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="add-progeny-pathway-scores" class="level2">
<h2 class="anchored" data-anchor-id="add-progeny-pathway-scores">3.3 Add PROGENy Pathway Scores</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load progeny package</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>library(progeny)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PROGENy pathway scores using myR function</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Attempting to add PROGENy scores...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>tryCatch({</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  seurat_norm <span class="op">&lt;-</span> myR::add_progeny_scores(</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    seurat_obj <span class="op">=</span> seurat_norm,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    organism <span class="op">=</span> <span class="st">"Human"</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    topn <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"PROGENy scores added to metadata successfully</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"- PROGENy pathways added:"</span>, length(grep(<span class="st">"^pathway_|^NFkB|^JAK-STAT"</span>, colnames(seurat_norm<span class="op">@</span>meta.data))), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>}, error <span class="op">=</span> function(e) {</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"Error adding PROGENy scores:"</span>, e$message, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"Skipping PROGENy scores for now. Continuing with analysis...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="custom-gene-set-signatures" class="level2">
<h2 class="anchored" data-anchor-id="custom-gene-set-signatures">3.4 Custom Gene Set Signatures</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define custom gene sets relevant to mIBD</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>barrier_genes <span class="op">&lt;-</span> c(<span class="st">"MUC2"</span>, <span class="st">"OCLN"</span>, <span class="st">"TJP1"</span>, <span class="st">"CLDN1"</span>, <span class="st">"CLDN2"</span>, <span class="st">"CLDN4"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>fibrosis_genes <span class="op">&lt;-</span> c(<span class="st">"COL1A1"</span>, <span class="st">"COL3A1"</span>, <span class="st">"TGFB1"</span>, <span class="st">"ACTA2"</span>, <span class="st">"FN1"</span>, <span class="st">"TIMP1"</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>stromal_genes <span class="op">&lt;-</span> c(<span class="st">"VIM"</span>, <span class="st">"PDGFRA"</span>, <span class="st">"PDGFRB"</span>, <span class="st">"FAP"</span>, <span class="st">"SMA"</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>gene_sets_custom <span class="op">&lt;-</span> <span class="bu">list</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  Barrier <span class="op">=</span> barrier_genes,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  Fibrosis <span class="op">=</span> fibrosis_genes,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  Stromal <span class="op">=</span> stromal_genes</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each custom gene set and add module score with error handling</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (set_name <span class="kw">in</span> names(gene_sets_custom)) {</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>  cat(paste(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Calculating score for:"</span>, set_name), <span class="st">"---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  gene_list_current <span class="op">&lt;-</span> gene_sets_custom[[set_name]]</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check which genes are available</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>  available_genes <span class="op">&lt;-</span> gene_list_current[gene_list_current <span class="op">%</span><span class="kw">in</span><span class="op">%</span> rownames(seurat_norm)]</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>  missing_genes <span class="op">&lt;-</span> setdiff(gene_list_current, available_genes)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"- Requested genes:"</span>, length(gene_list_current), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"- Available genes:"</span>, length(available_genes), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (length(missing_genes) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"- Missing genes:"</span>, paste(missing_genes, collapse <span class="op">=</span> <span class="st">", "</span>), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (length(available_genes) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Try AddModuleScore with error handling</span></span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>    tryCatch({</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>      seurat_norm <span class="op">&lt;-</span> AddModuleScore(</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>        seurat_norm,</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>        features <span class="op">=</span> <span class="bu">list</span>(available_genes),</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>        name <span class="op">=</span> set_name,</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>        assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Rename the module score column for consistency (AddModuleScore adds '1')</span></span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>      seurat_norm[[set_name]] <span class="op">&lt;-</span> seurat_norm[[paste0(set_name, <span class="st">"1"</span>)]]</span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>      seurat_norm[[paste0(set_name, <span class="st">"1"</span>)]] <span class="op">&lt;-</span> NULL <span class="co"># remove old column</span></span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>      cat(paste(set_name, <span class="st">"Score added successfully using AddModuleScore.</span><span class="ch">\n</span><span class="st">"</span>))</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>    }, error <span class="op">=</span> function(e) {</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a>      cat(<span class="st">"Error in AddModuleScore for"</span>, set_name, <span class="st">":"</span>, e$message, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a>      cat(<span class="st">"Falling back to manual calculation (mean expression).</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Alternative: Calculate score manually</span></span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a>      expr_data <span class="op">&lt;-</span> LayerData(seurat_norm, layer <span class="op">=</span> <span class="st">"data"</span>)</span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Use drop = FALSE to handle cases with a single gene without turning it into a vector</span></span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true" tabindex="-1"></a>      available_expr <span class="op">&lt;-</span> expr_data[available_genes, , drop <span class="op">=</span> FALSE]</span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-53"><a href="#cb14-53" aria-hidden="true" tabindex="-1"></a>      seurat_norm[[set_name]] <span class="op">&lt;-</span> colMeans(available_expr, na.rm <span class="op">=</span> TRUE)</span>
<span id="cb14-54"><a href="#cb14-54" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb14-55"><a href="#cb14-55" aria-hidden="true" tabindex="-1"></a>      cat(paste(set_name, <span class="st">"Score calculated manually.</span><span class="ch">\n</span><span class="st">"</span>))</span>
<span id="cb14-56"><a href="#cb14-56" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb14-57"><a href="#cb14-57" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-58"><a href="#cb14-58" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb14-59"><a href="#cb14-59" aria-hidden="true" tabindex="-1"></a>    warning(paste(<span class="st">"No genes for signature '"</span>, set_name, <span class="st">"' found in the dataset!"</span>))</span>
<span id="cb14-60"><a href="#cb14-60" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add a column of NAs as a placeholder</span></span>
<span id="cb14-61"><a href="#cb14-61" aria-hidden="true" tabindex="-1"></a>    seurat_norm[[set_name]] <span class="op">&lt;-</span> NA</span>
<span id="cb14-62"><a href="#cb14-62" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb14-63"><a href="#cb14-63" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-64"><a href="#cb14-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-65"><a href="#cb14-65" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">Custom signature scoring complete.</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="part-4-data-splitting-and-feature-engineering" class="level1">
<h1>Part 4: Data Splitting and Feature Engineering</h1>
<section id="metadata-feature-engineering" class="level2">
<h2 class="anchored" data-anchor-id="metadata-feature-engineering">4.1: Metadata Feature Engineering</h2>
<p>This section creates new metadata columns for analysis, including a unified <code>heal</code> score, <code>response</code> status, and various interaction terms. This is done <em>after</em> all signature scoring to prevent columns from being accidentally overwritten.</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the seurat_norm object which now has the complete metadata</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>md <span class="op">&lt;-</span> seurat_norm<span class="op">@</span>meta.data</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a unified 'heal' score based on a hierarchy of remission criteria</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>md <span class="op">&lt;-</span> md <span class="op">%&gt;%</span> mutate(</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  heal <span class="op">=</span> case_when(</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    `Deep mucosal healing` <span class="op">==</span> <span class="dv">1</span>                      <span class="op">~</span> <span class="dv">4</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    `Endo remission` <span class="op">==</span> <span class="dv">1</span>                             <span class="op">~</span> <span class="dv">3</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    `histologic Endoscopic improvement` <span class="op">==</span> <span class="dv">1</span>          <span class="op">~</span> <span class="dv">2</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    `mucosal healing` <span class="op">==</span> <span class="dv">1</span>                            <span class="op">~</span> <span class="dv">1</span>,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    `mucosal healing` <span class="op">==</span> <span class="dv">0</span>                            <span class="op">~</span> <span class="dv">0</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    <span class="kw">is</span>.na(`mucosal healing`)                          <span class="op">~</span> NA_real_,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    TRUE                                              <span class="op">~</span> NA_real_</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a binary 'response' status based on the heal score</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="co"># This is a different 'response' column than the one from the raw data</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>md$response_derived <span class="op">&lt;-</span> case_when(</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>  md[[VAR_RESPONSE_CRITERION]] <span class="op">==</span> <span class="dv">1</span> <span class="op">~</span> <span class="st">"Responder"</span>,</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>  md[[VAR_RESPONSE_CRITERION]] <span class="op">==</span> <span class="dv">0</span> <span class="op">~</span> <span class="st">"NonResponder"</span>,</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  TRUE <span class="op">~</span> NA_character_</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a simplified tissue type column</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>md$ck_ <span class="op">&lt;-</span> case_when(</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>    md$ck <span class="op">==</span> TRUE <span class="op">~</span> <span class="st">"ck"</span>,</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>    md$ck <span class="op">==</span> FALSE <span class="op">~</span> <span class="st">"str"</span>,</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>    TRUE <span class="op">~</span> <span class="st">"unknown"</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Create simplified binary response columns for each healing criterion</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>md$H1 <span class="op">&lt;-</span> ifelse(md$`mucosal healing` <span class="op">==</span> <span class="dv">1</span>, <span class="st">"R"</span>, <span class="st">"NR"</span>)</span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>md$H2 <span class="op">&lt;-</span> ifelse(md$`histologic Endoscopic improvement` <span class="op">==</span> <span class="dv">1</span>, <span class="st">"R"</span>, <span class="st">"NR"</span>)</span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>md$H3 <span class="op">&lt;-</span> ifelse(md$`Endo remission` <span class="op">==</span> <span class="dv">1</span>, <span class="st">"R"</span>, <span class="st">"NR"</span>)</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>md$H4 <span class="op">&lt;-</span> ifelse(md$`Deep mucosal healing` <span class="op">==</span> <span class="dv">1</span>, <span class="st">"R"</span>, <span class="st">"NR"</span>)</span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Create various interaction terms for downstream analysis and plotting</span></span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: We use 'treatment_original' to ensure we are using the raw 'pre'/'post' labels</span></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>md$r_t <span class="op">&lt;-</span> paste(md$response_derived, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>md$heal_t <span class="op">&lt;-</span> paste(md$heal, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>md$drug_r <span class="op">&lt;-</span> paste(md$drug, md$response_derived, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>md$drug_t <span class="op">&lt;-</span> paste(md$drug, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>md$drug_r_t <span class="op">&lt;-</span> paste(md$drug, md$response_derived, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>md$ck_r <span class="op">&lt;-</span> paste(md$ck_, md$response_derived, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>md$ck_t <span class="op">&lt;-</span> paste(md$ck_, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>md$ck_r_t <span class="op">&lt;-</span> paste(md$ck_, md$response_derived, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>md$drug_ck <span class="op">&lt;-</span> paste(md$drug, md$ck_, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>md$drug_ck_r <span class="op">&lt;-</span> paste(md$drug, md$ck_, md$response_derived, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>md$drug_ck_t <span class="op">&lt;-</span> paste(md$drug, md$ck_, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>md$drug_ck_r_t <span class="op">&lt;-</span> paste(md$drug, md$ck_, md$response_derived, md$treatment_original, sep <span class="op">=</span> <span class="st">"_"</span>)</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Update the Seurat object using AddMetaData, which handles row name alignment</span></span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>seurat_norm <span class="op">&lt;-</span> AddMetaData(seurat_norm, md)</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"--- Metadata Feature Engineering Complete ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"New columns added, including 'heal', 'response_derived', and various interaction terms.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(table(seurat_norm$heal, seurat_norm$response_derived, useNA <span class="op">=</span> <span class="st">"ifany"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="identify-pre-post-matched-pairs" class="level2">
<h2 class="anchored" data-anchor-id="identify-pre-post-matched-pairs">4.2: Identify Pre-Post Matched Pairs</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- New, More Robust Matching Logic using pre-defined pairs ---</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Identify valid pre/post pairs from the clinical metadata.</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># A valid pair has non-NA values for both key_pre and key_post.</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>valid_pairs <span class="op">&lt;-</span> mibd_meta_clinical <span class="op">%&gt;%</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(<span class="op">!</span><span class="kw">is</span>.na(key_pre) <span class="op">&amp;</span> <span class="op">!</span><span class="kw">is</span>.na(key_post))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Create a unique identifier for each pair, which will be used for the LMM random effect.</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>valid_pairs <span class="op">&lt;-</span> valid_pairs <span class="op">%&gt;%</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  mutate(pair_id <span class="op">=</span> paste(emrid, drugno, sep <span class="op">=</span> <span class="st">"_"</span>))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Get the list of all key_tma values that belong to these valid pairs.</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>paired_tma_keys <span class="op">&lt;-</span> unique(c(valid_pairs$key_pre, valid_pairs$key_post))</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Create the matched Seurat object by filtering the full dataset</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="co"># for cells whose key_tma is in our list of paired keys.</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>seurat_matched <span class="op">&lt;-</span> subset(seurat_norm, subset <span class="op">=</span> key_tma <span class="op">%</span><span class="kw">in</span><span class="op">%</span> paired_tma_keys)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. Add the 'pair_id' to the metadata of the matched object.</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="co"># This is crucial for running the paired statistical tests correctly.</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a><span class="co"># First, create a mapping from key_tma to pair_id.</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>key_to_pair_id_pre <span class="op">&lt;-</span> valid_pairs <span class="op">%&gt;%</span> select(key_tma <span class="op">=</span> key_pre, pair_id)</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>key_to_pair_id_post <span class="op">&lt;-</span> valid_pairs <span class="op">%&gt;%</span> select(key_tma <span class="op">=</span> key_post, pair_id)</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>key_to_pair_id_map <span class="op">&lt;-</span> bind_rows(key_to_pair_id_pre, key_to_pair_id_post) <span class="op">%&gt;%</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>  distinct()</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, join this map with the metadata of our matched Seurat object.</span></span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a><span class="co"># We use a temporary ID to preserve rownames during the join.</span></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>matched_meta <span class="op">&lt;-</span> seurat_matched<span class="op">@</span>meta.data</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>matched_meta$temp_id <span class="op">&lt;-</span> rownames(matched_meta)</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>matched_meta <span class="op">&lt;-</span> left_join(matched_meta, key_to_pair_id_map, by <span class="op">=</span> <span class="st">"key_tma"</span>)</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>rownames(matched_meta) <span class="op">&lt;-</span> matched_meta$temp_id</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>matched_meta$temp_id <span class="op">&lt;-</span> NULL</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>seurat_matched<span class="op">@</span>meta.data <span class="op">&lt;-</span> matched_meta</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Define the pre-only object (ALL pre-treatment samples) ---</span></span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>seurat_preonly <span class="op">&lt;-</span> subset(seurat_norm, subset <span class="op">=</span> treatment_original <span class="op">==</span> <span class="st">"pre"</span>)</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Report the results ---</span></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"--- Data Splitting Summary ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Total unique pre/post pairs identified from clinical data:"</span>, nrow(valid_pairs), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Total cells in the complete matched dataset (`seurat_matched`):"</span>, ncol(seurat_matched), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a>unmatched_cells <span class="op">&lt;-</span> <span class="bu">sum</span>(<span class="kw">is</span>.na(seurat_matched$pair_id))</span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(unmatched_cells <span class="op">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - WARNING:"</span>, unmatched_cells, <span class="st">"cells in seurat_matched could not be assigned to a pair_id!</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Total cells in the pre-only dataset (`seurat_preonly`):"</span>, ncol(seurat_preonly), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="create-analysis-subsets" class="level2">
<h2 class="anchored" data-anchor-id="create-analysis-subsets">4.3: Create Analysis Subsets</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by CK (or other segment identifier if applicable)</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="st">"ck"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(SP_final)) {</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  seurat_ck <span class="op">&lt;-</span> subset(seurat_norm, ck <span class="op">==</span> TRUE)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"CK subset:"</span>, ncol(seurat_ck), <span class="st">"samples</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by response status</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>seurat_responder <span class="op">&lt;-</span> subset(seurat_norm, response_status <span class="op">==</span> <span class="st">"Responder"</span>)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>seurat_nonresponder <span class="op">&lt;-</span> subset(seurat_norm, response_status <span class="op">==</span> <span class="st">"NonResponder"</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Response-based subsets:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Responders:"</span>, ncol(seurat_responder), <span class="st">"samples</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"- Non-responders:"</span>, ncol(seurat_nonresponder), <span class="st">"samples</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="verification-of-data-subsets" class="level2">
<h2 class="anchored" data-anchor-id="verification-of-data-subsets">4.4: Verification of Data Subsets</h2>
<p>This section provides summary tables to verify the contents of the <code>seurat_matched</code> and <code>seurat_preonly</code> objects. # It runs AFTER the filtering and splitting to give a clear picture of the final analysis cohorts.</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Helper function to generate a hierarchical summary table ---</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>generate_hierarchical_summary <span class="op">&lt;-</span> function(metadata, group_vars) {</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  summary_table <span class="op">&lt;-</span> metadata <span class="op">%&gt;%</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    mutate(patient_drug_pair <span class="op">=</span> paste(emrid, drug, sep <span class="op">=</span> <span class="st">"_"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    group_by(across(all_of(group_vars))) <span class="op">%&gt;%</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    summarise(</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>      `<span class="fl">1.</span> 환자<span class="op">&lt;</span>br<span class="op">&gt;</span>(emrid)` <span class="op">=</span> n_distinct(emrid),</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>      `<span class="fl">2.</span> 환자<span class="op">-</span>약물 쌍<span class="op">&lt;</span>br<span class="op">&gt;</span>(key)` <span class="op">=</span> n_distinct(patient_drug_pair),</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>      `<span class="fl">3.</span> 샘플<span class="op">&lt;</span>br<span class="op">&gt;</span>(patho_id)` <span class="op">=</span> n_distinct(patho_id),</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>      `<span class="fl">4.</span> ROI<span class="op">&lt;</span>br<span class="op">&gt;</span>(key_tma)` <span class="op">=</span> n_distinct(key_tma),</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>      `<span class="fl">5.</span> AOI<span class="op">&lt;</span>br<span class="op">&gt;</span>(SegmentDisplayName)` <span class="op">=</span> n(),</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>      .groups <span class="op">=</span> <span class="st">"drop"</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># escape=FALSE allows rendering of &lt;br&gt; tag in the header</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(knitr::kable(summary_table, caption <span class="op">=</span> <span class="st">"계층적 데이터 요약 (Hierarchical Data Summary)"</span>, escape <span class="op">=</span> FALSE))</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Overall Summary (Multi-drug patients included) ---</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"### 1. Overall Summary of Datasets</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Matched Pre/Post Dataset Summary ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>generate_hierarchical_summary(</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> seurat_matched<span class="op">@</span>meta.data,</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>  group_vars <span class="op">=</span> <span class="st">"treatment_original"</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Pre-Only (for R vs NR) Dataset Summary ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>preonly_meta_filtered <span class="op">&lt;-</span> seurat_preonly<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(<span class="op">!</span><span class="kw">is</span>.na(response_derived))</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>generate_hierarchical_summary(</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> preonly_meta_filtered,</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>  group_vars <span class="op">=</span> <span class="st">"response_derived"</span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Drug-Specific Summary ---</span></span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">### 2. Drug-Specific Breakdown</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Matched Pre/Post Data by Drug ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>generate_hierarchical_summary(</span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> seurat_matched<span class="op">@</span>meta.data,</span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>  group_vars <span class="op">=</span> c(<span class="st">"drug"</span>, <span class="st">"treatment_original"</span>)</span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Pre-Only (R vs NR) Data by Drug ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>generate_hierarchical_summary(</span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a>  metadata <span class="op">=</span> preonly_meta_filtered,</span>
<span id="cb18-52"><a href="#cb18-52" aria-hidden="true" tabindex="-1"></a>  group_vars <span class="op">=</span> c(<span class="st">"drug"</span>, <span class="st">"response_derived"</span>)</span>
<span id="cb18-53"><a href="#cb18-53" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="investigation-of-sample-discrepancies" class="level2">
<h2 class="anchored" data-anchor-id="investigation-of-sample-discrepancies">4.5: Investigation of Sample Discrepancies</h2>
<p>This section investigates why the final number of matched pairs for a drug might be lower than the number defined in the clinical metadata. This is typically due to one of the samples in a pair being filtered out during initial quality control.</p>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Forensic Investigation of the Missing Vedolizumab Pair ---</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n\n</span><span class="st">--- FORENSIC INVESTIGATION: The Missing Vedolizumab Pair ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Identify all defined Vedolizumab pairs from the original clinical metadata.</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>vedo_pairs_clinical <span class="op">&lt;-</span> mibd_meta_clinical <span class="op">%&gt;%</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(drug <span class="op">==</span> <span class="st">"Vedolizumab"</span>, <span class="op">!</span><span class="kw">is</span>.na(key_pre), <span class="op">!</span><span class="kw">is</span>.na(key_post))</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>cat(sprintf(<span class="st">"1. Clinical metadata defines </span><span class="sc">%d</span><span class="st"> pairs for Vedolizumab.</span><span class="ch">\n</span><span class="st">"</span>, nrow(vedo_pairs_clinical)))</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Extract all 'key_tma' values that SHOULD be in these pairs.</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>vedo_paired_keys_clinical <span class="op">&lt;-</span> unique(c(vedo_pairs_clinical$key_pre, vedo_pairs_clinical$key_post))</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>cat(sprintf(<span class="st">"   - These </span><span class="sc">%d</span><span class="st"> pairs correspond to </span><span class="sc">%d</span><span class="st"> unique key_tma values.</span><span class="ch">\n</span><span class="st">"</span>, nrow(vedo_pairs_clinical), length(vedo_paired_keys_clinical)))</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Get the list of 'key_tma' values that ACTUALLY EXIST in the final matched Seurat object.</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co">#    This is the crucial step: we check the object that is actually used for analysis.</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>keys_in_matched_seurat <span class="op">&lt;-</span> seurat_matched<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="bu">filter</span>(drug <span class="op">==</span> <span class="st">"Vedolizumab"</span>) <span class="op">%&gt;%</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>  pull(key_tma) <span class="op">%&gt;%</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>  unique()</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>cat(sprintf(<span class="st">"2. After QC and pairing logic, </span><span class="sc">%d</span><span class="st"> unique Vedolizumab key_tma values remain in the final `seurat_matched` object.</span><span class="ch">\n</span><span class="st">"</span>, length(keys_in_matched_seurat)))</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Find the missing key(s) by comparing the original list with the final list.</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>missing_keys <span class="op">&lt;-</span> setdiff(vedo_paired_keys_clinical, keys_in_matched_seurat)</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (length(missing_keys) <span class="op">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>  cat(sprintf(<span class="st">"3. DISCREPANCY: Found </span><span class="sc">%d</span><span class="st"> missing key_tma(s) in the final matched object: </span><span class="sc">%s</span><span class="ch">\n</span><span class="st">"</span>, length(missing_keys), paste(missing_keys, collapse <span class="op">=</span> <span class="st">", "</span>)))</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 5. Cross-reference the missing key(s) with the list of AOIs that were removed during QC.</span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>  <span class="co">#    'remove_aoi' (a vector of SegmentDisplayName) was created in the 'gene-filtering' chunk.</span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>  <span class="co">#    We use 'mibd_meta_raw' to find the 'key_tma' associated with the removed AOIs.</span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>  removed_aoi_details <span class="op">&lt;-</span> mibd_meta_raw <span class="op">%&gt;%</span></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(SegmentDisplayName <span class="op">%</span><span class="kw">in</span><span class="op">%</span> remove_aoi) <span class="op">%&gt;%</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>    select(SegmentDisplayName, key_tma)</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>  is_missing_key_in_qc_removed_list <span class="op">&lt;-</span> <span class="bu">any</span>(missing_keys <span class="op">%</span><span class="kw">in</span><span class="op">%</span> removed_aoi_details$key_tma)</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (is_missing_key_in_qc_removed_list) {</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"4. ROOT CAUSE IDENTIFIED:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"   - The sample(s) for the missing key(s) were REMOVED during the initial AOI quality control ('gene-filtering' chunk).</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"   - This happens when an AOI's total counts are too low (an outlier).</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Identify exactly which clinical pair is affected by this missing key.</span></span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>    affected_pair <span class="op">&lt;-</span> vedo_pairs_clinical <span class="op">%&gt;%</span></span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>      <span class="bu">filter</span>(key_pre <span class="op">%</span><span class="kw">in</span><span class="op">%</span> missing_keys <span class="op">|</span> key_post <span class="op">%</span><span class="kw">in</span><span class="op">%</span> missing_keys)</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"   - The specific patient-drug pair affected by this QC removal is:</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(affected_pair <span class="op">%&gt;%</span> select(emrid, drugno, key_pre, key_post))</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"</span><span class="ch">\n</span><span class="st">CONCLUSION: The Vedolizumab analysis correctly proceeds with 12 pairs because one of the original 13 pairs had a sample that failed quality control. Paired analysis requires both samples to be present.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"4. ROOT CAUSE UNKNOWN: The missing key was not found in the list of QC-removed AOIs. This requires further manual investigation.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"3. NO DISCREPANCY: All expected keys from the 13 pairs are present in the final matched object. If the count is still 12, the issue lies elsewhere.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="part-5-segmented-differential-expression-analysis" class="level1">
<h1>Part 5: Segmented Differential Expression Analysis</h1>
<section id="create-ck-based-subsets" class="level2">
<h2 class="anchored" data-anchor-id="create-ck-based-subsets">5.1 Create CK-based Subsets</h2>
</section>
</section>
<section id="now-we-split-the-complete-seurat_matched-object-into-ck-and-ck--subsets." class="level1">
<h1>Now, we split the complete <code>seurat_matched</code> object into CK+ and CK- subsets.</h1>
</section>
<section id="this-correctly-handles-cases-where-ck-status-might-differ-within-a-pair." class="level1">
<h1>This correctly handles cases where CK status might differ within a pair.</h1>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if 'ck' column is available</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="op">!</span><span class="st">"ck"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(seurat_norm<span class="op">@</span>meta.data)) {</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  stop(<span class="st">"Metadata column 'ck' not found! Cannot perform segmented analysis."</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsets for matched data</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>seurat_matched_ck_pos <span class="op">&lt;-</span> subset(seurat_matched, ck <span class="op">==</span> TRUE)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>seurat_matched_ck_neg <span class="op">&lt;-</span> subset(seurat_matched, ck <span class="op">==</span> FALSE)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsets for pre-only data</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>seurat_preonly_ck_pos <span class="op">&lt;-</span> subset(seurat_preonly, ck <span class="op">==</span> TRUE)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>seurat_preonly_ck_neg <span class="op">&lt;-</span> subset(seurat_preonly, ck <span class="op">==</span> FALSE)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"--- CK-based Subsets Created ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Matched CK+ samples:"</span>, ncol(seurat_matched_ck_pos), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Matched CK- samples:"</span>, ncol(seurat_matched_ck_neg), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Pre-only CK+ samples:"</span>, ncol(seurat_preonly_ck_pos), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Pre-only CK- samples:"</span>, ncol(seurat_preonly_ck_neg), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="initialize-structured-results-list-define-gene-list" class="level2">
<h2 class="anchored" data-anchor-id="initialize-structured-results-list-define-gene-list">5.2 Initialize Structured Results List &amp; Define Gene List</h2>
</section>
</section>
<section id="a-nested-list-to-hold-all-deg-results-and-a-list-of-genes-for-lmm." class="level1">
<h1>A nested list to hold all DEG results and a list of genes for LMM.</h1>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>deg_results <span class="op">&lt;-</span> <span class="bu">list</span>()</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Genes to test with LMM (computationally intensive, so we use a curated list)</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>candidate_genes <span class="op">&lt;-</span> unique(c(</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  inflammatory_genes, barrier_genes, fibrosis_genes, stromal_genes, <span class="st">"TNFRSF6B"</span>, <span class="st">"CSF3R"</span>, <span class="st">"CD84"</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="run-segmented-deg-analysis-loop" class="level2">
<h2 class="anchored" data-anchor-id="run-segmented-deg-analysis-loop">5.3 Run Segmented DEG Analysis Loop</h2>
</section>
</section>
<section id="this-loop-will-iterate-through-ck-and-ck--segments-and-run-all-comparisons." class="level1">
<h1>This loop will iterate through CK+ and CK- segments and run all comparisons.</h1>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reload the local myR package to ensure functions are available in this environment.</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"Reloading myR package to ensure all functions are available...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>devtools::load_all(<span class="st">"/data/kjc1/mylit/myR"</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the analysis configuration required by the LMM function</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co"># We now use 'pair_id' as the patient/subject identifier for a more accurate paired model.</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>my_config <span class="op">&lt;-</span> <span class="bu">list</span>(</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  timepoint <span class="op">=</span> <span class="st">"treatment_original"</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  drug <span class="op">=</span> <span class="st">"drug"</span>,</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>  response <span class="op">=</span> <span class="st">"response_status"</span>,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  patient <span class="op">=</span> <span class="st">"pair_id"</span>, <span class="co"># Use the new, more precise pairing ID</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  segment <span class="op">=</span> <span class="st">"ck"</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>analysis_segments <span class="op">&lt;-</span> <span class="bu">list</span>(</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>  CK_Positive <span class="op">=</span> <span class="bu">list</span>(</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>    matched <span class="op">=</span> seurat_matched_ck_pos,</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>    preonly <span class="op">=</span> seurat_preonly_ck_pos</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>  CK_Negative <span class="op">=</span> <span class="bu">list</span>(</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>    matched <span class="op">=</span> seurat_matched_ck_neg,</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>    preonly <span class="op">=</span> seurat_preonly_ck_neg</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (segment_name <span class="kw">in</span> names(analysis_segments)) {</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>  cat(paste(<span class="st">"</span><span class="ch">\n\n</span><span class="st">--- Starting Analysis for Segment:"</span>, segment_name), <span class="st">"---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 1. Pre vs. Post Analysis (Paired LMM) ---</span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- 1. Pre vs. Post Analysis (Paired LMM) ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a>  current_matched_obj <span class="op">&lt;-</span> analysis_segments[[segment_name]]$matched</span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ncol(current_matched_obj) <span class="op">&gt;</span> <span class="dv">0</span> <span class="op">&amp;&amp;</span> length(unique(current_matched_obj$treatment_original)) <span class="op">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a>    genes_for_lmm <span class="op">&lt;-</span> intersect(candidate_genes, rownames(current_matched_obj))</span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- DEBUGGING: Show a summary of the data being analyzed ---</span></span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"</span><span class="ch">\n</span><span class="st">    --- Debugging Info for Matched LMM ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(</span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a>      current_matched_obj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a>        group_by(drug, treatment_original) <span class="op">%&gt;%</span></span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a>        summarise(n_samples <span class="op">=</span> n(), .groups <span class="op">=</span> <span class="st">"drop"</span>)</span>
<span id="cb22-45"><a href="#cb22-45" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb22-46"><a href="#cb22-46" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"    --------------------------------------</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-47"><a href="#cb22-47" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-48"><a href="#cb22-48" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1.1) Total pre vs post</span></span>
<span id="cb22-49"><a href="#cb22-49" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - Running LMM for Total Pre vs. Post...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-50"><a href="#cb22-50" aria-hidden="true" tabindex="-1"></a>    lmm_total <span class="op">&lt;-</span> myR::run_lmm_multiple_genes(current_matched_obj, genes_for_lmm, my_config, n_cores <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb22-51"><a href="#cb22-51" aria-hidden="true" tabindex="-1"></a>    deg_results[[segment_name]][[<span class="st">'pre_vs_post'</span>]][[<span class="st">'total'</span>]] <span class="op">&lt;-</span> lmm_total$summary</span>
<span id="cb22-52"><a href="#cb22-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-53"><a href="#cb22-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1.2) By each drug</span></span>
<span id="cb22-54"><a href="#cb22-54" aria-hidden="true" tabindex="-1"></a>    deg_results[[segment_name]][[<span class="st">'pre_vs_post'</span>]][[<span class="st">'by_drug'</span>]] <span class="op">&lt;-</span> <span class="bu">list</span>()</span>
<span id="cb22-55"><a href="#cb22-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (drug_name <span class="kw">in</span> unique(current_matched_obj$drug)) {</span>
<span id="cb22-56"><a href="#cb22-56" aria-hidden="true" tabindex="-1"></a>      cat(paste(<span class="st">"  - Running LMM for Pre vs. Post, Drug:"</span>, drug_name), <span class="st">"...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-57"><a href="#cb22-57" aria-hidden="true" tabindex="-1"></a>      drug_subset <span class="op">&lt;-</span> subset(current_matched_obj, drug <span class="op">==</span> drug_name)</span>
<span id="cb22-58"><a href="#cb22-58" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb22-59"><a href="#cb22-59" aria-hidden="true" tabindex="-1"></a>      unique_treatments <span class="op">&lt;-</span> unique(drug_subset$treatment_original)</span>
<span id="cb22-60"><a href="#cb22-60" aria-hidden="true" tabindex="-1"></a>      cat(sprintf(<span class="st">"    - Drug: </span><span class="sc">%s</span><span class="st"> | Unique treatments found: </span><span class="sc">%s</span><span class="ch">\n</span><span class="st">"</span>, drug_name, paste(unique_treatments, collapse<span class="op">=</span><span class="st">", "</span>)))</span>
<span id="cb22-61"><a href="#cb22-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-62"><a href="#cb22-62" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(length(unique_treatments) <span class="op">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb22-63"><a href="#cb22-63" aria-hidden="true" tabindex="-1"></a>        cat(<span class="st">"    - Condition MET: Both pre and post samples found. Running LMM.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-64"><a href="#cb22-64" aria-hidden="true" tabindex="-1"></a>        lmm_drug <span class="op">&lt;-</span> myR::run_lmm_multiple_genes(drug_subset, genes_for_lmm, my_config, n_cores <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb22-65"><a href="#cb22-65" aria-hidden="true" tabindex="-1"></a>        deg_results[[segment_name]][[<span class="st">'pre_vs_post'</span>]][[<span class="st">'by_drug'</span>]][[drug_name]] <span class="op">&lt;-</span> lmm_drug$summary</span>
<span id="cb22-66"><a href="#cb22-66" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb22-67"><a href="#cb22-67" aria-hidden="true" tabindex="-1"></a>        cat(<span class="st">"    - Condition FAILED: Not enough treatment types (need both pre and post). Skipping.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-68"><a href="#cb22-68" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb22-69"><a href="#cb22-69" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb22-70"><a href="#cb22-70" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-71"><a href="#cb22-71" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1.3) By responder status</span></span>
<span id="cb22-72"><a href="#cb22-72" aria-hidden="true" tabindex="-1"></a>    deg_results[[segment_name]][[<span class="st">'pre_vs_post'</span>]][[<span class="st">'by_response'</span>]] <span class="op">&lt;-</span> <span class="bu">list</span>()</span>
<span id="cb22-73"><a href="#cb22-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (resp_status <span class="kw">in</span> c(<span class="st">"Responder"</span>, <span class="st">"NonResponder"</span>)) {</span>
<span id="cb22-74"><a href="#cb22-74" aria-hidden="true" tabindex="-1"></a>      cat(paste(<span class="st">"  - Running LMM for Pre vs. Post, Response:"</span>, resp_status), <span class="st">"...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-75"><a href="#cb22-75" aria-hidden="true" tabindex="-1"></a>      resp_subset <span class="op">&lt;-</span> subset(current_matched_obj, response_status <span class="op">==</span> resp_status)</span>
<span id="cb22-76"><a href="#cb22-76" aria-hidden="true" tabindex="-1"></a>       <span class="cf">if</span>(length(unique(resp_subset$treatment_original)) <span class="op">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb22-77"><a href="#cb22-77" aria-hidden="true" tabindex="-1"></a>        lmm_resp <span class="op">&lt;-</span> myR::run_lmm_multiple_genes(resp_subset, genes_for_lmm, my_config, n_cores <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb22-78"><a href="#cb22-78" aria-hidden="true" tabindex="-1"></a>        deg_results[[segment_name]][[<span class="st">'pre_vs_post'</span>]][[<span class="st">'by_response'</span>]][[resp_status]] <span class="op">&lt;-</span> lmm_resp$summary</span>
<span id="cb22-79"><a href="#cb22-79" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb22-80"><a href="#cb22-80" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb22-81"><a href="#cb22-81" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb22-82"><a href="#cb22-82" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - Skipping: No matched samples with both pre/post in this segment.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-83"><a href="#cb22-83" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb22-84"><a href="#cb22-84" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-85"><a href="#cb22-85" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 2. Non-Responder vs. Responder Analysis (Unpaired Wilcoxon) ---</span></span>
<span id="cb22-86"><a href="#cb22-86" aria-hidden="true" tabindex="-1"></a>  cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- 2. NR vs. R Analysis (Unpaired Wilcoxon on Pre-samples) ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-87"><a href="#cb22-87" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-88"><a href="#cb22-88" aria-hidden="true" tabindex="-1"></a>  current_preonly_obj <span class="op">&lt;-</span> analysis_segments[[segment_name]]$preonly</span>
<span id="cb22-89"><a href="#cb22-89" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-90"><a href="#cb22-90" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (ncol(current_preonly_obj) <span class="op">&gt;</span> <span class="dv">0</span> <span class="op">&amp;&amp;</span> length(unique(current_preonly_obj$response_status)) <span class="op">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb22-91"><a href="#cb22-91" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-92"><a href="#cb22-92" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 2.1) Total NR vs R</span></span>
<span id="cb22-93"><a href="#cb22-93" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - Running Wilcoxon for Total NR vs. R...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-94"><a href="#cb22-94" aria-hidden="true" tabindex="-1"></a>    deg_wilcox_total <span class="op">&lt;-</span> FindMarkers(</span>
<span id="cb22-95"><a href="#cb22-95" aria-hidden="true" tabindex="-1"></a>      current_preonly_obj, ident<span class="fl">.1</span> <span class="op">=</span> <span class="st">"Responder"</span>, ident<span class="fl">.2</span> <span class="op">=</span> <span class="st">"NonResponder"</span>,</span>
<span id="cb22-96"><a href="#cb22-96" aria-hidden="true" tabindex="-1"></a>      group.by <span class="op">=</span> <span class="st">"response_status"</span>, test.use <span class="op">=</span> <span class="st">"wilcox"</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb22-97"><a href="#cb22-97" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb22-98"><a href="#cb22-98" aria-hidden="true" tabindex="-1"></a>    deg_results[[segment_name]][[<span class="st">'nr_vs_r'</span>]][[<span class="st">'total'</span>]] <span class="op">&lt;-</span> deg_wilcox_total</span>
<span id="cb22-99"><a href="#cb22-99" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb22-100"><a href="#cb22-100" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 2.2) By each drug</span></span>
<span id="cb22-101"><a href="#cb22-101" aria-hidden="true" tabindex="-1"></a>    deg_results[[segment_name]][[<span class="st">'nr_vs_r'</span>]][[<span class="st">'by_drug'</span>]] <span class="op">&lt;-</span> <span class="bu">list</span>()</span>
<span id="cb22-102"><a href="#cb22-102" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (drug_name <span class="kw">in</span> unique(current_preonly_obj$drug)) {</span>
<span id="cb22-103"><a href="#cb22-103" aria-hidden="true" tabindex="-1"></a>      cat(paste(<span class="st">"  - Running Wilcoxon for NR vs. R, Drug:"</span>, drug_name), <span class="st">"...</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-104"><a href="#cb22-104" aria-hidden="true" tabindex="-1"></a>      drug_subset <span class="op">&lt;-</span> subset(current_preonly_obj, drug <span class="op">==</span> drug_name)</span>
<span id="cb22-105"><a href="#cb22-105" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (length(unique(drug_subset$response_status)) <span class="op">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb22-106"><a href="#cb22-106" aria-hidden="true" tabindex="-1"></a>        deg_wilcox_drug <span class="op">&lt;-</span> FindMarkers(</span>
<span id="cb22-107"><a href="#cb22-107" aria-hidden="true" tabindex="-1"></a>          drug_subset, ident<span class="fl">.1</span> <span class="op">=</span> <span class="st">"Responder"</span>, ident<span class="fl">.2</span> <span class="op">=</span> <span class="st">"NonResponder"</span>,</span>
<span id="cb22-108"><a href="#cb22-108" aria-hidden="true" tabindex="-1"></a>          group.by <span class="op">=</span> <span class="st">"response_status"</span>, test.use <span class="op">=</span> <span class="st">"wilcox"</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.1</span></span>
<span id="cb22-109"><a href="#cb22-109" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb22-110"><a href="#cb22-110" aria-hidden="true" tabindex="-1"></a>        deg_results[[segment_name]][[<span class="st">'nr_vs_r'</span>]][[<span class="st">'by_drug'</span>]][[drug_name]] <span class="op">&lt;-</span> deg_wilcox_drug</span>
<span id="cb22-111"><a href="#cb22-111" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb22-112"><a href="#cb22-112" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb22-113"><a href="#cb22-113" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb22-114"><a href="#cb22-114" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - Skipping: Not enough samples or response groups in this segment.</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb22-115"><a href="#cb22-115" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb22-116"><a href="#cb22-116" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-117"><a href="#cb22-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-118"><a href="#cb22-118" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n\n</span><span class="st">--- DEG Analysis Complete ---</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="part-6-structured-results-visualization" class="level1">
<h1>Part 6: Structured Results Visualization</h1>
<section id="helper-function-for-plotting" class="level2">
<h2 class="anchored" data-anchor-id="helper-function-for-plotting">6.1 Helper Function for Plotting</h2>
</section>
</section>
<section id="this-function-generates-a-volcano-plot-and-a-heatmap-for-a-given-deg-result." class="level1">
<h1>This function generates a volcano plot and a heatmap for a given DEG result.</h1>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>generate_deg_plots <span class="op">&lt;-</span> function(deg_df, title, seurat_obj_for_heatmap, annotation_cols) {</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="kw">is</span>.null(deg_df) <span class="op">||</span> nrow(deg_df) <span class="op">==</span> <span class="dv">0</span>) {</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    cat(paste(<span class="st">"No DEG results to plot for:"</span>, title, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>))</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>()</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add gene column if not present (from rownames)</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="st">"gene"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)) {</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    deg_df <span class="op">&lt;-</span> deg_df <span class="op">%&gt;%</span> mutate(gene <span class="op">=</span> rownames(.))</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Determine p-value and logFC columns (handling LMM vs Wilcoxon)</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"p_val_adj"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df) <span class="op">&amp;&amp;</span> <span class="st">"avg_log2FC"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)) {</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Wilcoxon result from FindMarkers</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>    p_val_col <span class="op">&lt;-</span> <span class="st">"p_val_adj"</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>    logfc_col <span class="op">&lt;-</span> <span class="st">"avg_log2FC"</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="st">"p.adj"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df) <span class="op">&amp;&amp;</span> <span class="st">"Estimate"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)) {</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># LMM result from custom function</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>    p_val_col <span class="op">&lt;-</span> <span class="st">"p.adj"</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>    logfc_col <span class="op">&lt;-</span> <span class="st">"Estimate"</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Fallback for unknown format, try original logic</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>    p_val_col <span class="op">&lt;-</span> <span class="cf">if</span> (<span class="st">"p.adj"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)) <span class="st">"p.adj"</span> <span class="cf">else</span> <span class="st">"p_val_adj"</span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>    logfc_col <span class="op">&lt;-</span> <span class="cf">if</span> (<span class="st">"avg_log2FC"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)) <span class="st">"avg_log2FC"</span> <span class="cf">else</span> <span class="st">"Estimate"</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="op">!</span>p_val_col <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df) <span class="op">||</span> <span class="op">!</span>logfc_col <span class="op">%</span><span class="kw">in</span><span class="op">%</span> colnames(deg_df)){</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>      cat(<span class="st">"Could not find p-value or logFC column for:"</span>, title, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">return</span>()</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Volcano Plot ---</span></span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a>  p_volcano <span class="op">&lt;-</span> ggplot(deg_df, aes_string(x <span class="op">=</span> logfc_col, y <span class="op">=</span> paste0(<span class="st">"-log10("</span>, p_val_col, <span class="st">")"</span>))) <span class="op">+</span></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>    geom_point(aes(color <span class="op">=</span> .data[[p_val_col]] <span class="op">&lt;</span> <span class="fl">0.05</span>), alpha <span class="op">=</span> <span class="fl">0.6</span>) <span class="op">+</span></span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>    geom_text_repel(</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a>      data <span class="op">=</span> deg_df <span class="op">%&gt;%</span> <span class="bu">filter</span>(.data[[p_val_col]] <span class="op">&lt;</span> <span class="fl">0.05</span>) <span class="op">%&gt;%</span> arrange(.data[[p_val_col]]) <span class="op">%&gt;%</span> head(<span class="dv">20</span>),</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>      aes(label <span class="op">=</span> gene), <span class="bu">max</span>.overlaps <span class="op">=</span> <span class="dv">20</span></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a>    ) <span class="op">+</span></span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>    labs(title <span class="op">=</span> paste(<span class="st">"Volcano Plot:"</span>, title), x <span class="op">=</span> <span class="st">"Log2 Fold Change"</span>) <span class="op">+</span></span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a>    theme_bw() <span class="op">+</span></span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a>    scale_color_manual(values <span class="op">=</span> c(<span class="st">"gray"</span>, <span class="st">"red"</span>), name <span class="op">=</span> <span class="st">"Significant (p.adj &lt; 0.05)"</span>)</span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(p_volcano)</span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Heatmap of Top DEGs ---</span></span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a>  top_degs <span class="op">&lt;-</span> deg_df <span class="op">%&gt;%</span></span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span>(.data[[p_val_col]] <span class="op">&lt;</span> <span class="fl">0.05</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-49"><a href="#cb23-49" aria-hidden="true" tabindex="-1"></a>    arrange(.data[[p_val_col]]) <span class="op">%&gt;%</span></span>
<span id="cb23-50"><a href="#cb23-50" aria-hidden="true" tabindex="-1"></a>    head(<span class="dv">50</span>) <span class="op">%&gt;%</span></span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a>    pull(gene)</span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (length(top_degs) <span class="op">&gt;</span> <span class="dv">0</span> <span class="op">&amp;&amp;</span> <span class="op">!</span><span class="kw">is</span>.null(seurat_obj_for_heatmap)) {</span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb23-55"><a href="#cb23-55" aria-hidden="true" tabindex="-1"></a>    pheatmap(</span>
<span id="cb23-56"><a href="#cb23-56" aria-hidden="true" tabindex="-1"></a>      LayerData(seurat_obj_for_heatmap, layer <span class="op">=</span> <span class="st">"data"</span>)[top_degs, ],</span>
<span id="cb23-57"><a href="#cb23-57" aria-hidden="true" tabindex="-1"></a>      annotation_col <span class="op">=</span> seurat_obj_for_heatmap<span class="op">@</span>meta.data[, annotation_cols, drop <span class="op">=</span> FALSE],</span>
<span id="cb23-58"><a href="#cb23-58" aria-hidden="true" tabindex="-1"></a>      show_colnames <span class="op">=</span> FALSE,</span>
<span id="cb23-59"><a href="#cb23-59" aria-hidden="true" tabindex="-1"></a>      main <span class="op">=</span> paste(<span class="st">"Top 50 DEGs:"</span>, title)</span>
<span id="cb23-60"><a href="#cb23-60" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb23-61"><a href="#cb23-61" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-62"><a href="#cb23-62" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualize-key-comparisons" class="level2">
<h2 class="anchored" data-anchor-id="visualize-key-comparisons">6.2 Visualize Key Comparisons</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Visualizing CK Positive Segment ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># NR vs R (Total)</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>generate_deg_plots(</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  deg_df <span class="op">=</span> deg_results$CK_Positive$nr_vs_r$total,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  title <span class="op">=</span> <span class="st">"CK+ | NR vs R (all drugs, pre-treatment)"</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  seurat_obj_for_heatmap <span class="op">=</span> analysis_segments$CK_Positive$preonly,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  annotation_cols <span class="op">=</span> c(<span class="st">"drug"</span>, <span class="st">"response_status"</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Visualizing CK Negative Segment ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># NR vs R (Total)</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>generate_deg_plots(</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  deg_df <span class="op">=</span> deg_results$CK_Negative$nr_vs_r$total,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  title <span class="op">=</span> <span class="st">"CK- | NR vs R (all drugs, pre-treatment)"</span>,</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  seurat_obj_for_heatmap <span class="op">=</span> analysis_segments$CK_Negative$preonly,</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  annotation_cols <span class="op">=</span> c(<span class="st">"drug"</span>, <span class="st">"response_status"</span>)</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="part-7-structured-summary-of-findings" class="level1">
<h1>Part 7: Structured Summary of Findings</h1>
<section id="summary-of-deg-results" class="level2">
<h2 class="anchored" data-anchor-id="summary-of-deg-results">7.1 Summary of DEG Results</h2>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>cat(<span class="st">"</span><span class="ch">\n\n</span><span class="st">--- SUMMARY OF ALL DEG COMPARISONS ---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (segment_name <span class="kw">in</span> names(deg_results)) {</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  cat(paste(<span class="st">"</span><span class="ch">\n</span><span class="st">--- Segment:"</span>, segment_name), <span class="st">"---</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pre vs Post (LMM)</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"pre_vs_post"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> names(deg_results[[segment_name]])) {</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - Pre vs Post (LMM on candidate genes):</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Total</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    total_df <span class="op">&lt;-</span> deg_results[[segment_name]]$pre_vs_post$total</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="op">!</span><span class="kw">is</span>.null(total_df)) {</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>      cat(sprintf(<span class="st">"    - Total: </span><span class="sc">%d</span><span class="st"> significant genes</span><span class="ch">\n</span><span class="st">"</span>, <span class="bu">sum</span>(total_df$p.adj <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> TRUE)))</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># By Drug</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(drug <span class="kw">in</span> names(deg_results[[segment_name]]$pre_vs_post$by_drug)) {</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>      drug_df <span class="op">&lt;-</span> deg_results[[segment_name]]$pre_vs_post$by_drug[[drug]]</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>      cat(sprintf(<span class="st">"    - Drug (</span><span class="sc">%s</span><span class="st">): </span><span class="sc">%d</span><span class="st"> significant genes</span><span class="ch">\n</span><span class="st">"</span>, drug, <span class="bu">sum</span>(drug_df$p.adj <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> TRUE)))</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># NR vs R (Wilcoxon)</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"nr_vs_r"</span> <span class="op">%</span><span class="kw">in</span><span class="op">%</span> names(deg_results[[segment_name]])) {</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>    cat(<span class="st">"  - NR vs R (Wilcoxon on all genes):</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Total</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    total_df <span class="op">&lt;-</span> deg_results[[segment_name]]$nr_vs_r$total</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="op">!</span><span class="kw">is</span>.null(total_df)) {</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>      cat(sprintf(<span class="st">"    - Total: </span><span class="sc">%d</span><span class="st"> significant genes</span><span class="ch">\n</span><span class="st">"</span>, <span class="bu">sum</span>(total_df$p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> TRUE)))</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># By Drug</span></span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(drug <span class="kw">in</span> names(deg_results[[segment_name]]$nr_vs_r$by_drug)) {</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>      drug_df <span class="op">&lt;-</span> deg_results[[segment_name]]$nr_vs_r$by_drug[[drug]]</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>      cat(sprintf(<span class="st">"    - Drug (</span><span class="sc">%s</span><span class="st">): </span><span class="sc">%d</span><span class="st"> significant genes</span><span class="ch">\n</span><span class="st">"</span>, drug, <span class="bu">sum</span>(drug_df$p_val_adj <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> TRUE)))</span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sessionInfo()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>